{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f13fe33e",
   "metadata": {},
   "source": [
    "![logo-gt.png](attachment:logo-gt.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e53587e1",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "    <h1>BIOS 4150/BIOL 6150</h1>\n",
    "    <h3>Instructor: Dr. King Jordan</h3>\n",
    "    <p>Shivam Sharma (shivamsharma13@gatech.edu), Nilavrah Sensarma (nsensarma3@gatech.edu), Bengy Gyimah-Asamoah, (bgyimaha3@gatech.edu)\n",
    "\n",
    "</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c342f3f",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h2>Project 3 (Variant Calling) starter notebook (100 points)</h2>\n",
    "    <h3>Deadline: 11:59PM, October 7th, 2024</h3>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3f25848",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <h2>Alert!</h2>\n",
    "    <p>Do not delete any comments or information boxes the starter notebook comes with</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "879a14ff",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f404ae11",
   "metadata": {},
   "source": [
    "# **1. Getting ready for variant calling**\n",
    "### *Total Questions: 6*\n",
    "### *Total Points: 5 + 10 + 5 + 20 + 5 + 20 = 65*\n",
    "\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ff09a1",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>1.1 Check your SAM file (you can also convert your SAM file to a BAM file and do the exercises on that)</h3>\n",
    "    <p>1.1.1 We will start from where we left off in Project #2. We want to make sure that we are using the corrent SAM file.</p>\n",
    "    <p>1.1.2 Show that your SAM file is sorted</p>\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "51df61f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check you SAM file here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d2a6a8f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Show that your SAM file is sorted by genomic coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6366633c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@HD\tVN:1.6\tSO:coordinate\r\n",
      "@SQ\tSN:chr1\tLN:248956422\r\n",
      "@SQ\tSN:chr10\tLN:133797422\r\n",
      "@SQ\tSN:chr11\tLN:135086622\r\n",
      "@SQ\tSN:chr11_KI270721v1_random\tLN:100316\r\n",
      "@SQ\tSN:chr12\tLN:133275309\r\n",
      "@SQ\tSN:chr13\tLN:114364328\r\n",
      "@SQ\tSN:chr14\tLN:107043718\r\n",
      "@SQ\tSN:chr14_GL000009v2_random\tLN:201709\r\n",
      "@SQ\tSN:chr14_GL000225v1_random\tLN:211173\r\n"
     ]
    }
   ],
   "source": [
    "#This was done in last project, but we want to check the sorted file again.\n",
    "!head ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaa21a7a",
   "metadata": {},
   "source": [
    "We know that the SAM file is sorted by genomic coordinates because the first line of the file (@HD) says coordinate after SO (sort order)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8943fedb",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>1.2 Basic statistics of your SAM file</h3>\n",
    "    <p>Get basic stats of your SAM/BAM file. Use SAMstats to see what summary information can you get from the tool. </p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a97fced",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: SAMstats [-h] --sorted_sam_file SORTED_SAM_FILE [--outf OUTF]\r\n",
      "                [--chunk_size CHUNK_SIZE]\r\n",
      "\r\n",
      "Compute SAM file mapping statistics for a SAM file sorted by read name\r\n",
      "\r\n",
      "options:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "  --sorted_sam_file SORTED_SAM_FILE\r\n",
      "                        Input SAM file. Use '-' if input is being piped from\r\n",
      "                        stdin. File must be sorted by read name.\r\n",
      "  --outf OUTF           Output file name to store alignment statistics. The\r\n",
      "                        statistics will be printed to stdout if no file is\r\n",
      "                        provided\r\n",
      "  --chunk_size CHUNK_SIZE\r\n",
      "                        Number of lines to read a time from sortedSamFile\r\n"
     ]
    }
   ],
   "source": [
    "# See what SAMStats can give you.\n",
    "!SAMstats --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90f33e55",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>1.3 The pieleup format</h3>\n",
    "    <p><b>1.3.1</b> A lot of variant callers use the pielup format for calling variants from SAM file. Explain in 2-3 lines what is the pileup format?</p>\n",
    "    <p><b>1.3.2</b> Explain the 6 columns of a pielup format file in your own words.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "725bd049",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Information about the pileup format."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aa28891",
   "metadata": {},
   "source": [
    "The pileup format is a readable text output so that humans can understand the information the VCF has about the chromosome variants. The format is generally as follows: Chromosome Position ReferenceBase NumberOfReads AlignedBases. The pileup format  uses characters that correspond to specific indentifiers. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "658cf70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Information about the 6 fields of the pieleup format file.\n",
    "#Write one sentence for each file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "951c2bcf",
   "metadata": {},
   "source": [
    "Column 1 is the sequence identifier, usually the chromosome. \n",
    "Column 2 is the position on the chromosome. \n",
    "Column 3 is the reference base at that position. \n",
    "Column 4 is the coverage depth, which is the number of reads for the chosen position. \n",
    "Column 5 is the read results, which are characterized by different symbols that correspond to various identifiers. This is usually a string of the coverage depth number of characters for each read at the chosen position. \n",
    "Column 6 is an optional column and it holds the Phred score."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de23fff0",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>1.4 Create the pileup file</h3>\n",
    "    <p>Using samtools, create the pileup file for the SAM file of your 1000 genomes individuals</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "741ddebd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 4.1G\r\n",
      "-rwxr-xr-x. 1 acressy3 gtperson 3.1G Oct  7 14:26 hg38.fa\r\n",
      "-rwxr-xr-x. 1 acressy3 gtperson 1.1G Oct  7 14:26 SRR393028.sorted.sam\r\n"
     ]
    }
   ],
   "source": [
    "#copy the files to project 3\n",
    "!mkdir -p ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling\n",
    "!cp ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling\n",
    "!cp ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling\n",
    "\n",
    "!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4ac166ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[mpileup] 1 samples in 1 input files\r\n"
     ]
    }
   ],
   "source": [
    "#Your command for pileup here.\n",
    "!samtools mpileup -f ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/hg38.fa ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.sorted.sam > ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "15c81b8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 10 lines\n",
      "chr1\t11132\tg\t1\t^!.\t?\n",
      "chr1\t11133\tt\t1\t.\t?\n",
      "chr1\t11134\ta\t1\t.\t?\n",
      "chr1\t11135\tg\t1\t.\t?\n",
      "chr1\t11136\ta\t1\t.\t?\n",
      "chr1\t11137\ta\t1\t.\t?\n",
      "chr1\t11138\tc\t1\t.\t?\n",
      "chr1\t11139\tc\t1\t.\t?\n",
      "chr1\t11140\tt\t1\t.\t?\n",
      "chr1\t11141\tc\t1\t.\t?\n",
      "Lines 10,000 - 10,010\n",
      "chr1\t485008\tT\t3\t...\t??]\n",
      "chr1\t485009\tC\t3\t...\t??]\n",
      "chr1\t485010\tA\t3\t...\t??]\n",
      "chr1\t485011\tG\t3\t...\t??]\n",
      "chr1\t485012\tA\t3\t...\t??]\n",
      "chr1\t485013\tA\t3\t...\t??]\n",
      "chr1\t485014\tT\t3\t..^!,\t?]]\n",
      "chr1\t485015\tA\t3\t..,\t?]]\n",
      "chr1\t485016\tG\t3\t..,\t?]]\n",
      "chr1\t485017\tT\t3\t..,\t?]]\n",
      "Lines 200,000 - 200,010\n",
      "chr1\t3823198\tA\t24\t,,,.,.,,..,,,.,,.,,,,,.,\t???]????]?]?????]??]????\n",
      "chr1\t3823199\tT\t24\t,,,.,.,,..,,,.,,.,,,,,.,\t???]????]?]?????]??]????\n",
      "chr1\t3823200\tG\t24\t,$,,.,.,,..,,,.,,.,,,,,.,\t???]????]?]?????]??]????\n",
      "chr1\t3823201\tA\t23\t,,.,.,,..,,,.,,.,,,,,.,\t??]????]?]?????]??]????\n",
      "chr1\t3823202\tC\t23\t,,A,.,,..,,,.,,.,,,,,.,\t??]????]?]?????]??]????\n",
      "chr1\t3823203\tA\t23\t,,.,.,,..,,,.,,.,,,,,.,\t??]????]?]?????]??]????\n",
      "chr1\t3823204\tC\t23\t,$,.,.,,..,,,.,,.,,,,,.,\t??]????]?]?????]??]????\n",
      "chr1\t3823205\tA\t22\t,$.,.,,.T,,,.,,.,,,,,.,\t?]????]?]?????]??]????\n",
      "chr1\t3823206\tG\t21\t.$,$.,,..,,,.,,.,,,,,.,\t\\>???]?]?????]??]????\n",
      "chr1\t3823207\tG\t20\t.,,..,,,.,,,.,,,,,.,\t???]?]??????]??]????\n"
     ]
    }
   ],
   "source": [
    "#Show the first 10 lines of the Pileup file.\n",
    "!echo 'First 10 lines'\n",
    "!head ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup\n",
    "#Show the 10,000 - 10,010 lines of the Pileup file.\n",
    "!echo 'Lines 10,000 - 10,010'\n",
    "!head ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup -n10010 | tail\n",
    "#Show the 200,000 - 200,010 lines of the Pileup file.\n",
    "!echo 'Lines 200,000 - 200,010'\n",
    "!head ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup -n200010 | tail"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49431713",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>1.5 Look more into the pileup file</h3>\n",
    "    <p>Show the 100,000th entry in your pileup file and explain (2-4 lines) what information column 5 <b><i>Read Results</i></b> is providing the user</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d8984c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t1638660\tG\t1\t,\t?\r\n"
     ]
    }
   ],
   "source": [
    "#Show the 100,000th entry here.\n",
    "!head ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup -n100001 | tail -n1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23218ee6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Explain the entry here."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b0536f0",
   "metadata": {},
   "source": [
    "The reference base is a G, indicated by column 3, and our read in column 5, denoted by a comma, means that the base is on the reverse strand of ther read but aligns to the reference genome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "367b7759",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>1.6 Plot the read count distribution</h3>\n",
    "    <p>A graph that shows how many reads are aligning to a specific genomic region could be helpful for determining flags for our variant caller. \n",
    "    <p>Plot a graph (thin bars or line) between <b><i>Read Count (y-axis)</i></b> and <b><i>Position (x-axis)</i></b> only for <u>chromosome 20</u>.</p>\n",
    "    <hr>\n",
    "    <p>You can use Python for visualization, but will have to filter the pileup file using bash or python.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8acd8c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#An awk command can help you create a file with just the information for chr20.\n",
    "!awk '/chr20/' ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup > ~/biol6150/ProjectSubmissions/Group8/Project3/chr20_reads.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "187b4640",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create CSV with only column 2 (Read Count) and 4 (Position)\n",
    "!awk '/chr20/ {print $2 \",\" $4}' ~/biol6150/ProjectSubmissions/Group8/Project3/chr20_reads.txt > ~/biol6150/ProjectSubmissions/Group8/Project3/chr20_reads.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "71aca26d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         position  read count\n",
      "0           87532           1\n",
      "1           87533           1\n",
      "2           87534           1\n",
      "3           87535           1\n",
      "4           87536           1\n",
      "...           ...         ...\n",
      "1636496    182188           1\n",
      "1636497    182189           1\n",
      "1636498    182190           1\n",
      "1636499    182191           1\n",
      "1636500    182192           1\n",
      "\n",
      "[1636501 rows x 2 columns]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAHFCAYAAAAHcXhbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABuXklEQVR4nO3deXxTZb4/8E/apmlLS4HShdICBcomW6EjiyI4CIrKiDo6o6jgdl1QRFRG5I7CqKBcxoszjPhzAx3XO+NyXUaFq04FAS1L2VWWslMKFdpC6Zrz++OYmKRZnnNytiSf9+uVFzR5kvPk5JyTb855vs/XJkmSBCIiIqIIFWd2B4iIiIjCwWCGiIiIIhqDGSIiIopoDGaIiIgoojGYISIioojGYIaIiIgiGoMZIiIiimgMZoiIiCiiMZghIiKiiMZghkgjy5cvh81mc98SEhKQl5eHm2++GYcPH9ZlmTabDXPnznX/vWPHDsydOxf79u1r1Xbq1Kno1q2bLv0w0759+7zWe1xcHNq3b4+xY8dixYoVpvWrW7dumDp1qlDbmpoaPPnkkyguLkbbtm3hcDjQrVs33HLLLdi4caO+HRUUbNsiMhuDGSKNLVu2DGvXrsXKlStx++2346233sKoUaNw5swZzZe1du1a3Hbbbe6/d+zYgXnz5vn9wvnjH/+I999/X/M+WMW9996LtWvXYtWqVVi0aBF27dqFSy+9FF9//bXZXQtqz549KCoqwlNPPYULL7wQb731FlasWIF58+bh2LFjGDp0KKqrq83uZtBti8hsCWZ3gCja9O/fH8XFxQCACy+8EC0tLXj88cfxwQcfYPLkyZoua/jw4cJte/TooemyraZLly7u9XHeeeehsLAQo0ePxssvv4wLLrjA5N7519LSgiuvvBInTpzA2rVr0b9/f/djo0ePxpQpU/Dpp5/Cbreb2Esi6+OZGSKdub5g9+/fDwCor6/H7NmzUVBQgMTERHTu3BnTpk3DqVOnvJ735ZdfYsyYMcjIyEBycjK6dOmCq6++GnV1de42npeZli9fjmuuuQaAHES5LrssX74cgP/LTKJ96datGy6//HJ89tlnGDJkCJKTk9GnTx+88sorQd97U1MTsrKycOONN7Z67NSpU0hOTsbMmTMBAE6nE0888QR69+6N5ORktGvXDgMHDsSzzz4bdBmBuALKY8eOed1fUVGBO+64A3l5eUhMTERBQQHmzZuH5uZmr3bz5s3DsGHD0KFDB7Rt2xZDhgzByy+/DN/avE1NTZg1axZycnKQkpKC888/H999951QHz/44ANs3boVs2fP9gpkPE2YMAEpKSnuv1evXo2xY8ciLS0NKSkpGDlyJD755BOv58ydOxc2m63Va7kuhXqeXRH5bENtW0Rm45kZIp3t3r0bAJCZmQlJkjBp0iR88cUXmD17NkaNGoUtW7bgsccew9q1a7F27Vo4HA7s27cPl112GUaNGoVXXnkF7dq1w+HDh/HZZ5+hsbHR68vN5bLLLsP8+fPxyCOP4G9/+xuGDBkCIPAZGdG+uGzevBkPPPAAHn74YWRnZ+Oll17Crbfeip49ewY882G323HDDTfg+eefx9/+9je0bdvW/dhbb72F+vp63HzzzQCAhQsXYu7cufjP//xPXHDBBWhqasL333/fKrASVV5eDgDo1auX+76Kigqce+65iIuLw6OPPooePXpg7dq1eOKJJ7Bv3z4sW7bM3Xbfvn2444470KVLFwDAunXrcO+99+Lw4cN49NFH3e1uv/12vPbaa3jwwQcxbtw4bNu2DVdddRVqa2tD9tE1pmfSpElC76mkpATjxo3DwIED8fLLL8PhcOC5557DxIkT8dZbb+F3v/ud0Ov4CvXZKt22iAwnEZEmli1bJgGQ1q1bJzU1NUm1tbXSxx9/LGVmZkppaWlSRUWF9Nlnn0kApIULF3o995133pEASC+88IIkSZL0z3/+UwIglZWVBV0mAOmxxx5z//2Pf/xDAiB99dVXrdpOmTJF6tq1q/tv0b5IkiR17dpVSkpKkvbv3+++7+zZs1KHDh2kO+64I2gft2zZ0ur1JEmSzj33XGno0KHuvy+//HJp8ODBQV/Ln/LycgmA9PTTT0tNTU1SfX29VFZWJo0YMULq1KmTVF5e7m57xx13SKmpqV7vQ5IkadGiRRIAafv27X6X0dLSIjU1NUl/+tOfpIyMDMnpdEqSJEk7d+6UAEj333+/V/s33nhDAiBNmTIlaN8vueQSCYBUX18v9F6HDx8uZWVlSbW1te77mpubpf79+0t5eXnufj322GOSv8O7axv1XCein22wbYvIbLzMRKSx4cOHw263Iy0tDZdffjlycnLw6aefIjs7G19++SUAtMpyueaaa9CmTRt88cUXAIDBgwcjMTER//Ef/4FXX30Ve/fu1byfon1xGTx4sPssBQAkJSWhV69e7stngQwYMABDhw71Ouuxc+dOfPfdd7jlllvc95177rnYvHkz7r77bnz++eeoqalR9H7+8Ic/wG63IykpCYMHD8a2bdvw0UcfeV1a+/jjj3HhhRciNzcXzc3N7tuECRMAyGc+XL788ktcdNFFSE9PR3x8POx2Ox599FFUVVWhsrISAPDVV18BQKuxUNdeey0SErQ98X3mzBl8++23+O1vf4vU1FT3/fHx8bjxxhtx6NAh/PDDD6peW+1nS2QVDGaINPbaa6+htLQUmzZtwpEjR7Blyxacd955AICqqiokJCQgMzPT6zk2mw05OTmoqqoCIJ++/7//+z9kZWVh2rRp6NGjB3r06KF6/Ig/on1xycjIaPUaDocDZ8+eDbmsW265BWvXrsX3338PQM74cjgcuO6669xtZs+ejUWLFmHdunWYMGECMjIyMHbsWKxfv17o/dx3330oLS3F6tWrsWjRIjQ1NeGKK67weh/Hjh3DRx99BLvd7nU755xzAAAnTpwAAHz33XcYP348AODFF1/EN998g9LSUsyZMwcA3O/Z9do5OTlefUlISPC7vny5AgjXJbFgTp48CUmS0KlTp1aP5ebmevVHqXA+WyIrYDBDpLG+ffuiuLgYgwcPbvXFk5GRgebmZhw/ftzrfkmSUFFRgY4dO7rvGzVqFD766CNUV1dj3bp1GDFiBGbMmIG3335bk34q6Uu4rrvuOjgcDixfvhwtLS34+9//jkmTJqF9+/buNgkJCZg5cyY2btyIn376CW+99RYOHjyIiy++2GvQcyB5eXkoLi7GeeedhwceeAAvvfQSDh8+jMcee8zdpmPHjhg/fjxKS0v93m699VYAwNtvvw273Y6PP/4Y1157LUaOHOkeUOzJFQRUVFR43d/c3CwUWFx88cUA5IHAobRv3x5xcXE4evRoq8eOHDnifn+AfGYFABoaGrzauYI1omjDYIbIQGPHjgUAvP766173v/vuuzhz5oz7cU/x8fEYNmwY/va3vwFA0EnUXAN2RX5Rq+mLWu3bt8ekSZPw2muv4eOPP0ZFRYXXJSZf7dq1w29/+1tMmzYNP/30k6q5TSZPnowxY8bgxRdfdF8uufzyy7Ft2zb06NEDxcXFrW6uMxyuSQ/j4+Pdr3f27Fn8/e9/91rGmDFjAABvvPGG1/3/8z//0yo7yp8rrrgCAwYMwIIFC7Bt2za/bT7//HPU1dWhTZs2GDZsGN577z2vz9fpdOL1119HXl6ee7Cz69Lali1bvF7ro48+CtmnQJRsW0RGYzYTkYHGjRuHiy++GH/4wx9QU1OD8847z51BVFRU5E5hfv755/Hll1/isssuQ5cuXVBfX+9Olb3ooosCvr4rvfeFF15AWloakpKSUFBQ4PcygmhftHLLLbfgnXfewT333IO8vLxW72PixInuOXoyMzOxf/9+LF68GF27dkVhYaGqZT799NMYNmwYHn/8cbz00kv405/+hJUrV2LkyJGYPn06evfujfr6euzbtw//+te/8PzzzyMvLw+XXXYZnnnmGVx//fX4j//4D1RVVWHRokVe2V2AfBbuhhtuwOLFi2G323HRRRdh27ZtWLRokVfmViDx8fF4//33MX78eIwYMQJ33XUXLrzwQrRp0wb79+/HP//5T3z00Uc4efIkAGDBggUYN24cLrzwQjz44INITEzEc889h23btuGtt95yp2Nfeuml6NChA2699Vb86U9/QkJCApYvX46DBw+qWo+Asm2LyHAmD0AmihquTJHS0tKg7c6ePSv94Q9/kLp27SrZ7XapU6dO0l133SWdPHnS3Wbt2rXSlVdeKXXt2lVyOBxSRkaGNHr0aOnDDz/0ei34ZDNJkiQtXrxYKigokOLj4yUA0rJlyyRJap3NJNoXSZIzXi677LJW72X06NHS6NGjg75fl5aWFik/P18CIM2ZM6fV43/+85+lkSNHSh07dpQSExOlLl26SLfeequ0b9++oK/rymb6r//6L7+PX3PNNVJCQoK0e/duSZIk6fjx49L06dOlgoICyW63Sx06dJCGDh0qzZkzRzp9+rT7ea+88orUu3dvyeFwSN27d5cWLFggvfzyy62ygRoaGqQHHnhAysrKkpKSkqThw4dLa9eulbp27Roym8nl1KlT0uOPPy4NGTJESk1Nlex2u9SlSxfphhtukL755huvtqtWrZJ+/etfS23atJGSk5Ol4cOHSx999FGr1/zuu++kkSNHSm3atJE6d+4sPfbYY9JLL73kN5tJ9LMNtG0Rmc0mST4zQBERERFFEI6ZISIioojGYIaIiIgiGoMZIiIiimgMZoiIiCiiMZghIiKiiMZghoiIiCJa1E+a53Q6ceTIEaSlpbknlCIiIiJrkyQJtbW1yM3NRVxc8HMvUR/MHDlyBPn5+WZ3g4iIiFQ4ePAg8vLygraJ+mAmLS0NgLwyRKYXJyIiIvPV1NQgPz/f/T0eTNQHM65LS23btmUwQ0REFGFEhohwADARERFFNAYzREREFNEYzBAREVFEYzBDREREEY3BDBEREUU0BjNEREQU0RjMEBERUURjMENEREQRjcEMERERRbSonwGYiKKEJAHVtUBjE5BoB9LTABaPJSIwmCGiSHD8JLD7gBzIuCTagZ5dgMz25vWLiCyBl5mIyNqOnwR27PEOZAD57x175MeJKKYxmCEi65Ik+YxMMHsOyO2IKGYxmCEi63KNkQmmoUluR0Qxi8EMEVlXqEBGaTsiikoMZojIuhLt2rYjoqjEYIaIrCs9LXSg4vg5TZuIYhaDGSKyLpsNiAsxl4zNxvlmiGIcgxkisq7mZqC+MXib+ka5HRHFLAYzRGRd3+/Vth0RRSUGM0RkXWdDnJVR2o6IohKDGSKyruREbdsRUVRiMENE1tWnu7btiCgqMZghIutKEKyFK9qOiKISgxkisq7qam3bEVFUYjBDRNZVtkvbdkQUlRjMEBERUURjMENEREQRjcEMEVnX4EJt2xFRVGIwQ0TWlZ6ubTsiikoMZojI2uwh0q5DPU5EUY/BDBFZV2Mj0BSiiGRTs9yOiGIWgxkisq7NP2jbjoiiEoMZIrKuhhBnZZS2I6KoxGCGiKzLITgeRrQdEUUlU4OZuXPnwmazed1ycnLcj0uShLlz5yI3NxfJyckYM2YMtm/fbmKPichQg3pr246IopLpZ2bOOeccHD161H3bunWr+7GFCxfimWeewZIlS1BaWoqcnByMGzcOtbW1JvaYiAyTmKhtOyKKSqYHMwkJCcjJyXHfMjMzAchnZRYvXow5c+bgqquuQv/+/fHqq6+irq4Ob775psm9JiJDnD2rbTsiikqmBzO7du1Cbm4uCgoK8Pvf/x579+4FAJSXl6OiogLjx493t3U4HBg9ejTWrFkT8PUaGhpQU1PjdSOiCLVhp7btiCgqmRrMDBs2DK+99ho+//xzvPjii6ioqMDIkSNRVVWFiooKAEB2drbXc7Kzs92P+bNgwQKkp6e7b/n5+bq+ByLSUYtT23ZEFJVMDWYmTJiAq6++GgMGDMBFF12ETz75BADw6quvutvYbDav50iS1Oo+T7Nnz0Z1dbX7dvDgQX06T0T6ixc8RIm2I6KoZKkjQJs2bTBgwADs2rXLndXkexamsrKy1dkaTw6HA23btvW6EVGEGtpX23ZEFJUsFcw0NDRg586d6NSpEwoKCpCTk4OVK1e6H29sbERJSQlGjhxpYi+JyDDJydq2I6KoZOpMUw8++CAmTpyILl26oLKyEk888QRqamowZcoU2Gw2zJgxA/Pnz0dhYSEKCwsxf/58pKSk4Prrrzez20RkFNGaS42NTM8mimGmBjOHDh3CddddhxMnTiAzMxPDhw/HunXr0LVrVwDArFmzcPbsWdx99904efIkhg0bhhUrViAtLc3MbhORUZTUZvrVAH37QkSWZZMkSTK7E3qqqalBeno6qqurOX6GKNKs3gS0tIRuFx8PnF+kf3+IyDBKvr8tNWaGiMgLazMRkQAGM0RkXazNREQCGMwQkXUlJgJxgeeVAiA/zsG/RDGNwQwRWZfTCThDDOtzSnI7IopZDGaIyLqOVGrbjoiiEoMZIrKusw3atiOiqMRghoisK9mhbTsiikoMZojIunKztG1HRFGJwQwRWVdcHJCaErxNaorcjohiFo8ARGRdTidwui54m9N1zGYiinEMZojIupjNREQCGMwQkXUxm4mIBDCYISLrYjYTEQlgMENE1sVsJiISwGCGiKwrLg7Iyw7eJi+b2UxEMY5HACKytlO14T1ORFGPwQwRWVdzs1hqdnOzMf0hIktiMENE1vX9Xm3bEVFUYjBDRNZ1tlHbdkQUlRLM7gAREQBAkoDqWqCxCUi0A+lpQHIiUFcf+rnJifr3j4gsi8EMEZnv+Elg9wE5kHFJtAPdcoGqmtDP79Ndv74RkeUxmCEicx0/CezY0/r+xibgx/2ADYAU5PlxNiCBhzKiWMYxM0RkHkmSz8gEbRPiNZwSs5mIYhyDGSIyj2uMTLiYzUQU0xjMEJF5tAhkAGYzEcU4BjNEZJ5Euzavw2wmopjGYIaIzJOepk1Aw2wmopjGYIaIzGOzAT27hP86zGYiimkMZojIXJntgX49Wt/vsAN5mWKvUSMwFw0RRS0GM0Rkvsz2QEL8L38P6gUMGwgcOi72/E0/6tMvIooIPDdLRNZgs/3y/3ZtzesHEUUcnpkhIuNIEnCqBqiskv+VPGbEa/KY+G71RuDsWeP7R0QRiWdmiMgYgeov9ewC7PQpZ9DiBL7bLv7aRb206SMRRSQGM0Skv2D1l/zdr1RbXpYiimW8zERE+hKpvxQu1mYiimkMZohIX1rVXwqGtZmIYhqDGSLSl96BDMDaTEQxjsEMEelLq/pLwbA2E1FMYzBDRPrSqv5SMKzNRBTTGMwQkb60qr8USGoKazMRxTgGM0SkPz0H6J6p0++1iSgiMJghIn3V1wNOKXQ7tSRwtmCiGMdghoj0tXGn/svYYMAyiMiyGMwQkb6aWvRfRotT/2UQkWUxmCEifdnj9V9GPA9lRLGMRwAi0teQvvovY6gByyAiy2IwQ0T6SkrSfxnJyfovg4gsi8EMEemroUH/ZTSynAFRLGMwQ0T62vS9/svY/IP+yyAiy2IwQ0T6amzWfxkNBiyDiCyLwQwR6SvRgFIDDpYzIIpllglmFixYAJvNhhkzZrjvkyQJc+fORW5uLpKTkzFmzBhs377dvE4SkXJFffRfxqDe+i+DiCzLEsFMaWkpXnjhBQwcONDr/oULF+KZZ57BkiVLUFpaipycHIwbNw61tbUm9ZSIFHM49H39OBuQmKjvMojI0kwPZk6fPo3JkyfjxRdfRPv27d33S5KExYsXY86cObjqqqvQv39/vPrqq6irq8Obb75pYo+JSJGfftL39Z0S4OQMwESxzPRgZtq0abjssstw0UUXed1fXl6OiooKjB8/3n2fw+HA6NGjsWbNmoCv19DQgJqaGq8bEZloq44Vs12OVOq/DCKyLFNHzb399tvYuHEjSktLWz1WUVEBAMjOzva6Pzs7G/v37w/4mgsWLMC8efO07SgRWdtZA+ayISLLMu3MzMGDB3Hffffh9ddfR1KQGUJtNpvX35IktbrP0+zZs1FdXe2+HTx4ULM+E5FFJes8LoeILM20YGbDhg2orKzE0KFDkZCQgISEBJSUlOAvf/kLEhIS3GdkXGdoXCorK1udrfHkcDjQtm1brxsRmWhAd/2XkZul/zKIyLJMC2bGjh2LrVu3oqyszH0rLi7G5MmTUVZWhu7duyMnJwcrV650P6exsRElJSUYOXKkWd0mIqU6dND39ZMSgTjTh/8RkYlMGzOTlpaG/v37e93Xpk0bZGRkuO+fMWMG5s+fj8LCQhQWFmL+/PlISUnB9ddfb0aXiUiN06f1ff36RqClBYiP13c5RGRZlp42c9asWTh79izuvvtunDx5EsOGDcOKFSuQlpZmdteISNQGA2oz7T0IFHbTfzlEZEk2SZIkszuhp5qaGqSnp6O6uprjZ4jMULJe/2W0S+MswERRRsn3t6XPzBARCUn5OZuprg7YsFOeSC/OBgztC6SkmNs3omjT0iKfDa1rkPe97vmmX+ZlMENE+hraR/9LTd3zW58BckpA6Q75/6OL9V0+UazYtguoqv7l71O1wJETQEY60L/QtG4xBYCI9JWaqu/r2wCs3hS8jRGXuoiinW8g46mqWn7cJAxmiEhfZ87o+/qio/7q6nTtBlFUa2kJHMi4VFXL7UzAYIaI9LVhp9k9kFmlH0SRaK/gbPqi7TTGYIaI9GWVfEmnVTpCFIHqBOufibbTGIMZItJX4FJqxoqzSkeIIlCKYP0z0XYaYzBDRPoa2tfsHsis0g8S09IC7NoHbP5B/teksRj0s+752rbTGFOziUhfbdqE9/w4W/BLRPYEoKk59OtwvpnIYdH035gWHy+v/2CDgDPSTZtvhmdmiMjaQh0cHYnG9IOMYeH0X7IuBjNEpK9wU7NDnXU5LZhyzdRs67N4+m9Ms/hnw2CGiPRllZRoq/SDArN4+m9Ms/hnw2CGiPRllYxopmZbn8XTf2OaxT8bBjNEpC+rZEQzNdv6LJ7+G9Ms/tkwmCEifVklJdoq/aDALJ7+G9Ms/tkwmCEifYWbmh2KXXCGCaZmW198fOgzaHE209J/Y5roOmdqNhFFpVOn9H19kTlmAKCpSd9+UPiamkKPbXJK/CzNUF+vbTuNMZghIn1t3m12D2Scn8T6RD8jfpbG2yiYDSjaTmMMZogoNtTz17zliX5G/CyN1yQ4f4xoO40xmCGi2JBkN7sHFIroZ8TP0nh2wbEwou00xmCGiPQ1qKfZPZCxpo/1iX5G/CyNN0QwG1C0ncYYzBCRvtq10/f187KB+BCHsvg4wM5f85Yn+hnxszReUpK27TTGYIaI9NUsmG2kVudMoMUZvE2LE3CGaEPmE81SYjaT8U6e1LadxhjMEJG+vt+r7+uLZk8cqdS3HxQ+ZjNZ15Y92rbTGIMZItLX2UZ9X180e+Is6/lYHrOZSCXBqTOJyK2hAdj0PdDYDCQmAEV9AAdrxQSUnAjU6TiRlj1eLKBJjtLPSJKA6lqgsQlItAPpaYAtQutQJdnl9yHSLtY0NclnpOqb5Pffv5BjhzwwmCFSYvVG7/EZDU3Auq3yANPzh5jXLyvr0x34pky/1x/SF/h2W+h2uVn69cEsx08Cuw94BwCJdqBnFyCzvXn9Uqt/IbBms1i7WPLtFqDe4wxnY5O8npISgWEDjenDwB5il5AG9tC/L37wMhORKN9AxlOLU36cWkvQ+TfT4eNi9Xziouxwd/wksGNP6zMZjU3y/cfNGYgZFrs9dJV1G2LrjIRvIOOpvlF+3AjtBYNj0XYai7K9m0gnDQ1iGTMNHJfRSqPOY2YOHROr56N3P4wkSfIZmWD2HJDbRZLmZiBUlyXonyFnFU1NgQMZl/pGY7K7DoTY3pS20xiDGSIRm77Xtl0s2fyD2T2QWaUfWnCNkQmmoUluF0lEM9/0zpCzCitld5ULZgOKttMYgxkiEY2CvwRF28WSBousE6v0Qwsig2SVtLMK0cw3vTPkrILZXcIYzBCJSBQc9yHaLpY4LLJOrNIPLSQKjhkRbWcVyYnatot0rFUljMEMkYiiPtq2iyWDepvdA5lV+qGF9LTQgYoNQOVPQIs5VYxV6dNd23aRzkq1qgoEswFF22mMwQyRCIdDrP4P55tpLTERsOt4ViQvO/Tr2xPkfkQLm01Ovw5GAnD0BLB6U+TMmJuQAKSmBG+TmqJ/hpxV2O1y+nUwSYnGZHd1CbG9KW2nMQYzRKJEspnIv+wM/V77VIQNctXKTgXTxldVR05AEypTKVYymVxCBW6xEtiFoDiYee2119DgJ/20sbERr732miadIrKcujpt28USp1NOn9bL6TqgKcQXXFNzdKVmnz0bOoXZV1W19S85WSkV2Qqam+XtO5jTdcYEeBUV2rbTmOJg5uabb0Z1dXWr+2tra3HzzTdr0ikiy9kgWMxQtF0ssUqBx2hKzVa7ne09qG0/tGalVGQrsFKq+g+HtG2nMcXBjCRJsPmp+3Ho0CGkp6dr0ikiywk1KZvSdrHEKgUeoyk1W+0lzTqLfBaBMBXZG1PVhQlfbCsqKoLNZoPNZsPYsWOR4HGdrqWlBeXl5bjkkkt06SSR6eJsYoFKqGn1raalRf61XtcApDiA7vlAfLy2y7BKgcdgqdmRVjw0Pk5dQJNi4fcE6FtoMhILcooWadU6Vd3f/mBxwsHMpEmTAABlZWW4+OKLkZqa6n4sMTER3bp1w9VXX615B4ksYWhfoHSHWLtIsW2XPI7C5VQtcOQEkJGubapnbhawx5xTz14CpWZHYvHQoX2B77Yrf173fO37oiW9Ck1GakFO0SKtWqaqB9ofRPXO064vCggHM4899hgAoFu3bvjd736HpKQk3TpFZDkpIdJFlbYzm28g48mV+aJVQBMXJ6dP6zUIOM4m/+oOduLMBv+p2SLFQ60Y0CQnq3ue1mfdtCaaYqwkFdlVkNOXqyBnvx7WDWhEM5W0ymgKtj+IysnRpi8KKV4DU6ZMASBnL1VWVsLp9H7jXUzKMSfSlWgByYYGa1+eAORLS4ECGRdX5otWX349fj4joEdAI3L5T0Lr96OkeKjVPtMzZ9Q9r6YGaNtW275oSfR9nTkDtGkTup1oQc6O7ax5yamqSrxdRpjTH4jsD0b1RQXFA4B37dqFUaNGITk5GV27dkVBQQEKCgrQrVs3FBQU6NFHIvNFU6FJ0YwWrTNfeuR7/wLukSeftTGK7/uJ5M9UbTbTph+17YfWtM4ajPSCnNvKtW0XjFbbuRZ9UUHxmZmpU6ciISEBH3/8MTp16uQ3s4ko6kRToUnRjBY9Ml88g5e8HGPH0vi+n0j+TKM1aU70fYm2i9aCnHqw4naugOJgpqysDBs2bECfPtYf3UykmcQE+RecSDurS3GIzZqrR+aL56SCqzaIZ4lpwff9RPJnaoM5AY3e2W+i70v0N3S0FuTUQ0JCRE9GqPgcb79+/XDixAk9+kJkXdFUaFI0o0XrzJeS9UDt2V/+dkrGzsvj+34i+TNVmzVX1Ev9Mrftkus8HTnxS+ab1nWfRN+XaDuRgpyOn9O0rai/4NAN0XaBfLtFu0Am3L6opDiYefrppzFr1iz8+9//RlVVFWpqarxuRFHJ4Qj9a9AG6w0U9Uf0l7SWv7hL1mv3Wr5Ef6X7vh/Rz8qKn6nI4Fd/1A7+Fcl+04Lo+xJtZ7MBaQKFK606XEJ0IG04A26/3RK6hIQSJgz+BVRcZrrooosAAGPHjvW63zUzcIvVa38QqeF0hj79Lf3czshBrWoYnZmld70q0ZM7tbVAmscv8JMnxZ538iTQ3mKpu6J999XUpLzCspHZb8cEs92OHQOys0O3czrF+m7V/VbvbVSkFpZSlZVAVpa2rylAcTDz1Vdf6dEPImsTrS90pFIe2GplSrJ4hg8Kf3lWqVe18QdgdPEvf28RrDq9ZY/386xAtO++tu0Civope46S7LfCboq75OV7wWV9f1AsmIn0/VbvbVSPGlc7D0RGMDN69Gg9+kFkbaL1haxShygYo7N4WK/KOtTUNDIz+y1c0bTf6iGKalwpDma+/vrroI9fcMEFqjtDZFmi9YWsUocoGKOzeIzMWKLg1NQ0MjP7LVzRtN/qQbQWVgRQfJFwzJgxrW4XXnih+6bE0qVLMXDgQLRt2xZt27bFiBEj8Omnn7oflyQJc+fORW5uLpKTkzFmzBhs366iHglRuHIFT5uKtjOT0Vk8VqlXNcSnNtPAHmLPE21nJLV9UlOiwsjstz6CryHaLtL3W723US1rsLn0NacKgOJg5uTJk163yspKfPbZZ/jVr36FFStWKHqtvLw8PPXUU1i/fj3Wr1+PX//617jiiivcAcvChQvxzDPPYMmSJSgtLUVOTg7GjRuH2lqLztZI0SsuDkgKUZk2KdGagwh9GZ3Fo3e9KtFElDSf9FvRAZNWG/wLqOuTDcoH/wLGZr+JjINR0k50f7Tqfqv3Nmq3A3aN51EyYbwMoCKYSU9P97p17NgR48aNw8KFCzFr1ixFrzVx4kRceuml6NWrF3r16oUnn3wSqampWLduHSRJwuLFizFnzhxcddVV6N+/P1599VXU1dXhzTffVNptovC0tIQe9V/fKLezukbB7AXRdqFYJZvJN1Pm9Gmx54m2M5Ka2kwSgGYV46CU1AcK1yHBGaFF2ymp9WRF1SEysZS28yVJ2qelm7S/aBaOZmZm4ocfflD9/JaWFrz99ts4c+YMRowYgfLyclRUVGD8+PHuNg6HA6NHj8aaNWsCvk5DQwPnviHtmVXPSA+bBfdT0XahWCWbyTdTZoNgVpdoOyOpXaff71X+HCPrA+2p0Lad1rWejFYmmG0k2s6XSO0qpUzaXxSfX9qyZYvX35Ik4ejRo3jqqacwaJDyNM6tW7dixIgRqK+vR2pqKt5//33069fPHbBk+5xOzM7Oxv79+wO+3oIFCzBv3jzF/SAKKpIzOnw1CP46F20XCgf/ak/tKj2r8ZwiVqd1radoEyWDfwEVwczgwYNhs9kgSd6f/vDhw/HKK68o7kDv3r1RVlaGU6dO4d1338WUKVNQUlLifty3kKVrcr5AZs+ejZkzZ7r/rqmpQX6+xtOy6625Wf4FdbYRSE4E+nSX62ZYldMpz9NwtkHOCsjNsu41aLUiOaPDlyMBqBO4HOZgNpPb1xvk7K6iPtqNJTp92vtX7NA+QGqq2HPV1mZKDjHuK9poXesp2kRRTSrFR6vycu9TiXFxccjMzERSUpKqDiQmJqJnz54AgOLiYpSWluLZZ5/FH/7wBwBARUUFOnXq5G5fWVnZ6myNJ4fDAYcVpx8XtWEHcNpjjEFdPfBNmTzl9lCFk10ZYc9B4JDPWIQ9h4C8bKBHhAWRwXTPl2vRiLSzukG9gbVbxNppYWhfoHSHNq8VDt8MmKF9xE+JS5Kczr5uKxAfB5w/JLy++Cvv4OqLyORnQ/sC61VcGunTXflz+heIXULSoiZPjxyxS0g9BCe4E11PVsm48zW4UOwS0mCVWUmu2lVanqEZak4tM8U/n7t27ep1y8/PVx3I+CNJEhoaGlBQUICcnBysXLnS/VhjYyNKSkowcuRIzZZnKb6BjKfTdfLjVuIvkHE5dEx+PFrEx4vVZtKynpFeEgV/nYu2C0XvbCbRbAzfH0GiZ0F8tTiB1RvVPRcIXadKpI6V2tpMas7wGlEfyCUvT6zdMcGxkFrXejJaerq27XyJ1K5SSu1+FSZV1wJKSkowceJE9OzZE4WFhfjNb36DVatWKX6dRx55BKtWrcK+ffuwdetWzJkzB//+978xefJk2Gw2zJgxA/Pnz8f777+Pbdu2YerUqUhJScH111+vptvW1twcOJBxOV2nLhtBD05n4EDG5dAxuV00aGwUq82kVQaQnozO8JAkfU9nNwnuEz/95P13OPtSi1O8xpUnrTKo1GbNqemzkduLaFaO6LFQSR0yKxKtZK224rVI7SqlTPqOUhzMvP7667jooouQkpKC6dOn45577kFycjLGjh2rOGX62LFjuPHGG9G7d2+MHTsW3377LT777DOMGzcOADBr1izMmDEDd999N4qLi3H48GGsWLECab7zRUQD0SwDNdkIelBS8yQaGJ0BpCejMzz0yJhQY6vPvhPuviRa48qTVhlUarPmVPXZwO1FSVaOyOenpA6ZFYnWTlJbY0mP47NJ31GKzzk++eSTWLhwIe6//373fffddx+eeeYZPP7444rOmrz88stBH7fZbJg7dy7mzp2rtJuRRzTLwCrZCLFW88ToDCA9GZ3hYYVAxp9w9yWtalepoTZrTk2frZoRJPL5GV2HTGuitZPU1ljS4/hs0neU4jMze/fuxcSJE1vd/5vf/KbV4GBSQDTLwCrZCLFW80Q0s0erDCA9iWZuaJXhYdWMiXD3Ja1qV6mhNmtOTZ+N3l5EiXx+ou/XzM8yGNFaWmpqbgH6HJ9N+o5SHMzk5+fjiy++aHX/F198EXkp0FYimmWgJhtBD5Fe80Qp0cwerTKA9CSauaFVhocrY8JsA3z2nXD3JaW1q+rrxduGyghRmzXn2eeWFmDXPvnS6K59gcfhGLW9KB33IfL5GV2HTA1JAk7VAJVV8r+e056I1k5SW2NJj+OzSd9RisPRBx54ANOnT0dZWRlGjhwJm82G1atXY/ny5Xj22Wf16GNsSEiQ06+DDQJOTbHOfDNxcXL6dbBBwHnZ0TPfTGKinDUTbLCpPUG7DCA9GZ3hYbMBPbsAO/Zo83q+Qm2HLh06eP8tss8FEh+nbL6ZVRuUzbUTKiMkPh7ISFc2eNOzz9t2eT/3VK089UBGeusvRiO2l2+3hC4X4kn0WOhwyO+7JUgigtLPUkvHTwK7D3hfik20y/tLZnu5dlJSYvB1k5SoruYWIHYc9xRq3h4Tv6MUL/Wuu+5CTk4O/vznP+N//ud/AAB9+/bFO++8gyuuuELzDsYUkWwmKzlxMvTj0TTXjCMxeDDjiIBAxix6BTKAPIgx1Bd7RoDUVbX7VG8Fc6ooDWRENSg8k+H6QvcNZDxVVcuP61FNORClgQxgvWOhGsdP+t8vGpvk+/v1kAMakZpw4Th6XLxtmxDBf5dOgR/TmaoQ6sorr8SVV16pdV9im5L0RyvMidDUJLaTNTWp/9VgJUpS561y9iyQkyGCUM92WlSM9k2J1ppTCn2GoqpavoziOQ9QOAUw9xwAOrYLXaSvvl5dIBNq3Ytsj/5UVytfV7UCM1+72inNNBU5jgRSVxd6DqOGhuBnZYBf0uyNPDsjSfIZmWD2HABSBH8giawLf0TWj6dQ25zofqED4WsAJ0+exF//+le/hRurq6sDPkaCIq0gmt4pg1YTaanzwWwRPEsi2i4U35Ros/imM4ezLzU0ySnnoWxUuYxQ617tdiaa+uy5rjYKTjcg2s5TOMcHkc/PqqnZItMVNDTp/72g9fsW3S90IBzMLFmyBF9//TXatm3b6rH09HSsWrUKf/3rXzXtXEyxavpjIHqnDFpNpKXOU2u+6czhXvoRSTlvUjm5XSh6b2dGFUwN5/gg8vlZNTVbdLoC0U1U7basx/s2aSoG4WDm3XffxZ133hnw8TvuuAP//Oc/NelUxBDNBhARbvqj0wkcqgB27Zf/1XvmXb1TBq0m0lLnqTXfdOa4ME+F7z4gF6BctznwDLJ2ncpb6L2dGVUwNZzjg8jnZ9XUbNHsPtFNVO22rMf7NilzUTiY2bNnDwoLAw8KKywsxJ49Og7ys5ptu4DVm+QMAFcmwOpN6k+bhpP+uOcgsGqjXODxyHH531Ub9a2NpHfKoNVEWup8MAN7aNsuFN+UaLP4pjOHnUrc4l2A0l+9piEqlxFq3avdzkQLEnquqyGC0w2ItvMUzvFB5POzamq2yHQFDrv+afFav2+HXX5vJhAOZuLj43HkyJGAjx85cgRx0ZKGG4pINoBSatMfzSr2KDqoNxoG/wLig3qtPvgXEB/Uq8XgX6B1SrQeQv0yjbO1LgKqdQFMfwUo1RbhDbXu1W5n6elyKm8oa8p++b/ooF41ZWbCOT6IfH6ig3qNTs222cTGzIh+L6jdlrV+3zabKYN/AQXBTFFRET744IOAj7///vsoKirSok/W1tIing2gxKlTytuZWexR7wJoVnP2rLbtzCQ6UF+rAf2ixQPDEWrMgFNqvS2K7nNK+BagVFvA8HiIdNlDh9S9bn29WPaQU/plkj/R4qlqiqxWVSl/jotIVp7RRVVFiR4nRPdBtcdZrY9X9Y3WLzR5zz334M9//jOWLFmCFo8v6paWFvz1r3/Ff//3f2PatGm6dNJSRAu8KS0Et3m38nZmFnuMtWymSMs2C2bTj9q2C0VJ8UA9+W6LovucUp4ZImqzRXbsD/74ngp1r6sku8rVVs8iq9vCKIEjkm1n1f1WdHmi+6Da46we79vqhSavvvpqzJo1C9OnT8ecOXPQvXt32Gw27NmzB6dPn8ZDDz2E3/72t3r21RpER/kbkQ1gZrHHWMtmEp2LQcmcDWQso7ZFzwwRqxUwVJJd5WobyUVWrZolqvVxQu22rcfxyqSMTkUXXp988klcccUVeOONN7B7925IkoQLLrgA119/Pc4991y9+mgtKQ55wK9IO724JrQys9hjkl0sBS9asplCTYnu2Y6syaht0TNDJDFB+Uy9erLHiwc0rkwsRwJQJ/Cclp8HRJs0ZsKvUNPve7YzkujxRJTabVvrfgCmZXTaJEmyyswluqipqUF6ejqqq6v9zpGjWEuLnLUUyvlFrQccBnPqlLLT3hnp8nTXq/xkUPgaNUT7GklNTcCazaHbjRwUHYOAz54Fvtseut255wDJyfr3Jxw1NWKnr4t6AVrsM9XV1rjU5LstKt3nRA0f8MvAyoYGOdNJqX5dgczMwI8fOqTuUtOw/sC328TbJiXJY2HWbhFfhmsafhFVVeovNQ3sEXqg9JkzwHqBSynFfY2dWV30eFLUS2xfVXucFe2HEucN1iwRQsn3N39GKiUaoCgJZACgXTtl7auq5fodoX5R2KBPscdYy2YSDVCsHsgA4gGKFoEMIGfQ6M0e4uBpT2i9LSrd50R5ZoiozRYJFsgAQF6e8teMs8nBSapg5osrE0tp8dQde+S6QyIyMpS9tieRbDuji6qKSk4WO3aL7oNqj7N6HK9MyuhkMKPUiRPatnOpUPErq6o69ClUCeozKoKJtWwmq2ZFqCG6PWi13RiR4RXq0kacTb4E4km05pBSnhkoajI7HPbWffUlWl/Lk1NSVtPJ9fmryVLatT/0ewB+yZhSQ6S2ltHbuhIXFAcOaGw/P65nJhkQXn2yQPTar0JgMKPU9n3atnP5QWWqpQg96o4wmym8dmYyul6NEetEZM4O35oxamoJifC8LKAms0Okvo3aullK+uP6/NVkKTU161u7Cojs2kwuFxQDmR5nLuPj5EvVFxTLf+uZSQbos2/qtV+FEAEzfFHY9MioiLVsJqtmRahhdL0aq2R4mVEzRm1mh159VdIf1+evNktJ79pVkVybyVNyMoCf52I6f4j3Y3pnkoVbn8xCeGbGbE1NwKYd+i5Dj/obsVabKdzaWVZidL0aq2R4mVEzRm1mh159VdIf1+fvULkdhBrHBIRXuyqSazN58rwcd6rG+2/Rda/2Mwq3PpmFCGUzFRUVwSaYbrdxo0B2jYE0z2Y6cULsEtI53YCOHYO3+XaL2Gyc4fLMrtBKrGUzWTUrQg3RDButths9MiZ8JYaYKsBhB4YN9B5bU1ur3ynxvGygR748RuWbMmXP9ddXXydPqrvUdN5g8f64Pn+l2UwuiXagZ5fgmU319eLZVb5+1S/0NP5Gb+tKHT8J/FDuffbSc72JrvsRA5UP1AbkMTOlGv+YHtJbXWkLP5R8fwuFc5MmTXL/v76+Hs899xz69euHESNGAADWrVuH7du34+6771bf60jRsSOAfYLtgtAikBGdI0CPnTTWspmsmhWhhtH1avTO8IqPA7I6BC/tkdmhdXCg0QHXL1dfeuQHb+eP3R56QLOaulnxccoyTVyfv5ovSUAOLnfsCZ6qrbZ2FRDZtZkAOZDZ4ScgFVlvvtR+RlrXJwP03a+CENqyH3vsMff/b7vtNkyfPh2PP/54qzYHD+pYpdkqRLMTmpsDHziamrQ5IyM6FuHkSe2KBrqI1tuprjYmNVdvSrK3rB7AGf1e9M5uaHEClT8Fb3P8J6B7nneQoCaDUIlDx4AcFUU2T9f9MjFmIGrWaYsTOBailpunEyfkH2Xh1ujacwDo2M5/gBZOFtHZs6EDZSV1yLSaikCEJAG7DwRvs+cAkC7440jtvqpH9uWpU/pNexCE4ovZ//jHP3DTTTe1uv+GG27Au+++q0mnLE00GyBYO6MzfNRmPgQjOgmaFSZL00I0ZW8Z/V6MyG5Qk82kZwahi9pskVC13dSu0+8V/OB0XU4Pt0ZXsOyscLKIhLKZDK5DJqq6VmybFc1SslJtJr1qnoWgOJhJTk7G6tWrW92/evVqJIVzyjBSiGYDBGsXLRk+sSSasrei6b0oYUY2k9pkESNquxkp0LoPJ4vIKllyaohui6JZSmr31ehJZlKemj1jxgzcdddd2LBhA4YPHw5AHjPzyiuv4NFHH9W8g5aTnAjUCUz0FCxrQLSuEVlHNNWiiqb3ooQZ2UyitYF86VnbzQyB1n04tauskiWnhui2KFoXS+2+qnb7tCDFW8PDDz+M1157DZs2bcL06dMxffp0bNq0CcuXL8fDDz+sRx+tpU/38Nv1L9SmL6IG9tD+NQcLvgfRdlYn+pkZ/dmqYfR7GdJbm9cJJtSXQ3wckOYz/qC3ipIASg3tq+553UMMHFa7TvsoGJB8Tjf536Je6pbl4rAD6QEGhRb1Uf+6Q/sChw8DJet/uR0+7PP6gn0P9z0qlZ4Wept12IFBgp9zz3xgzSZ5HazZJD6zstrtM5hBPbV/TQGqktOvvfZaXHvttVr3JTIkJMi1TYJNCZ6aEjxrwG4HkhLDHwQcqh8uWg/+BcQH9UbD4F9A7DNLSrT+4F/A+Peid3ZDRjqQ3dF/ZohLi1MuEOtKmQaAnBx9x83kZcvZbUp//Wakh67tpmadJiUC2dni42ZcGZnhDoxtbAqcneVwqKvcbIP/dP/dR+Xb6J9n0DW6Dpkom01Ovw62zfboImcpiXxXeI6hamqR093jbMCoocGfp0f2pQmDfwFOmqdOqABCJMDQIptJTd0XLYkU94smbUJkToR63EpC/fIP9bjVZLaXU1lDFVU9dAzYY1DWpSs9W+lpfLWXXUJRkpat5SUcCcCqDYEfV3OcCLVOS9b/8v9+Ic5Mh3pcL8ECGc/H1Y4NckrB13uUUbzFtrS0YNGiRTj33HORk5ODDh06eN2inpJUv0BOn9amL6IBkZqidKE0Nsr1V4JpalZfAM1qWlrkwp7BVFXL7axONC1UpFCgCL1ToF3rPbO9nAIcyqFjgNMJHAixDrSwS0WWyem60D9UqqrUve7x42JtW5y/pE1rkb7rlPxf+tBqmgp/Dh82flsXJfod8NNPoY+zwQRa7y6nTql/7UBEp+3QmOJgZt68eXjmmWdw7bXXorq6GjNnzsRVV12FuLg4zJ07V4cuWowWqX4bDC5qpkdqtt4F0KwmVKqs0nZmEk0LFSkUKMKIFGjXej8rmAV0pBIor9SvP+7lqDywh5oCYlu5utfdsV+8rSttWqv0XX9FJfWcymD3UeO3dVGi3wFbVRQq9RWsmKceadQmTcehOJh544038OKLL+LBBx9EQkICrrvuOrz00kt49NFHsW7dOj36SFakdwE0qxFNlY2ElFrRTLpIyrhzrXfRM2OiQY9Z1Bao1JIrbVqrkxb+ikrqnf4fjdu6UuEU84wgioOZiooKDBgwAACQmpqK6p9PKV1++eX45JNPtO1dpFMzujxSiBY2S4yXi6dVVrUuohZJRFNlzU6pdTqBQxXArv3yv04/19tF00LNSGVWy7XeQ42ZcUm2eOqz2gKVWtK6+KK/opJ6p/+LnnGJpG1dqXCKeUYQxcFMXl4ejh49CgDo2bMnVqxYAQAoLS2Fw4z6FkZTksLniohdo8tdg7GGhpGOqIYeqdmiKYMtErD5R2Bnufzvui1yTZJIIzog1syBs3sOAqs2AnsOAUeOy/+u2th6wKtoWmigdFqljEiBtsXJ25XI4HsAyM0CCrL07RMA5KrM5gs1BUT/AnWv26+reNuiPtoOlh7iJw1Y76kMjpwI3UbLbV2U6HfAAMGpQILxt95d9EijNmk6DsXBzJVXXokvvvgCAHDffffhj3/8IwoLC3HTTTfhlltu0byDlhNOCp9rdHlqqnb9EaFHarZoYTPf07euImqRFtDEx8spksEkJYZOqdXLnoOBCy36ZvDYbKHrMwVLp1UqJ0eb1wnmcGXo7BAXe4J8BqdLF337BACFKg/soTKPMjKUv6YNQGamePtDlcGLdyrlb4Z4K0xlYLNpt62LEv0O0CKpJtjM/HqkUZs0HYfi84hPPfWU+/+//e1vkZeXhzVr1qBnz574zW9+o2nnLCnc7BynJI9QN1J9fXjVaf0RLVYYSLDic1bkdIbOuqhvlNuJXurQitMZ+kvn0DGgoLPct/r60OMgJGi33Rw5Ev5raKmpWc4WMuLSr9pskVBFFM+eVf6aEoDvFSQfaBnIAEBdXesqzXoXIRVR3xi8MLAeRDP89isYsB1IbW3geYlEs9uU+HY9MKxY+9cNIeyj7vDhwzFz5szYCGQAbbJztBihrkSw0exqhZuFYEYGQTiOCGa+iLbTktK+iW4PWm03uywWzABytpARWYVqs0VCZRCpzTA6ptG0EGr467MRRUhFiBYQ1opoht8+DYKNYOtYSXabKJOGh6oKZv7+97/jvPPOQ25uLvb/HDkuXrwY//u//6tp5ywpErNz9BjNrkUWQiRlEIhmv5iRJaO0b6LbQzRnQVghWyiYUBOlRWKRRaeFB/9bfXugkBQHM0uXLsXMmTNx6aWX4tSpU2j5ORWyXbt2WLx4sdb9sx7RLB4r0WM0uxZZCKIZBM3NwLYfgdJt8r9mzHwsmv0i2q6hAVi3Gfh6g/xvQxhBkNK+iW4P0ZwFYYVsoWBCzcAbiUUW4yx8Sdnq2wOFpHiP+Otf/4oXX3wRc+bMQbzHYMfi4mJs3bpV085ZkmgWTzBajFBXIthodrXCzUIQzSDYsAP4pgyoqpGrlVfVyH9v2BHe8pXKFcx8EWm3eiOwbqt8qU2S5H/XbZXvN6JvotuDVttNYa42r6OlPt2NzypUIlQBQLUFArMNTj7w5K/PRhQhFSFaQFgrohl+3RQM2A4k2DpWkt0mSuPhmaIUBzPl5eUoKipqdb/D4cAZLaa9tjrRLJ5gjC77oPXgX0DOQgjnh1aDQLbMhh2BU21P1xkb0IgO6g3VbvXGwJcIWpzqAhqlfRPdHrTabnItGMwkJBifVahEsMG/Io8H0sfEAM538C+gfxFSUUYO/gXEM/y6ahBsBFvHSrLbRJkw+BdQEcwUFBSgrKys1f2ffvop+vXrp0WfrE2Lukqi9Z20ckzjrARAvtQT7iXwYOuyuVmsoKdRl5xEM1+CtWtoEBsLofSSU6XgAGBXO9EfHVr9OFFTn0hvdQrqFBnNnhB6cskTAvOn+LPD4DOanur87M961AZSw+hJTUU/v4MazPPjb7276FFHqVxlqY0wKQ5mHnroIUybNg3vvPMOJEnCd999hyeffBKPPPIIHnroIT36aC1aZECI1nfSyvc61AvSYvR/sHUp+vpGZSFokQG0SXDbEW3nslOwYKKrnWgmjFY1edTUJ2rr51e8ljbs1CeTQwtNzaEz/bbvU/faxwUnFdSDv+1Jj9pAauiR8RmM6Oe3V4MfosH2Yz3qKB1QUQRVA4rPrd18881obm7GrFmzUFdXh+uvvx6dO3fGs88+i9///vd69JGsSO/R/6Kvb1QWghYZQI2CZ5FE26klekbNzOSTztlAjY6/8KycWQNEVqafKCuv82jO3LPyeteQqguFt99+O26//XacOHECTqcTWVnywMLDhw+jc+fOmnaQNFK6TR6x36e7NteHkxPlAbl6EX19o7IQ7PFiB7xgGUCJCfJYoVC0romjlpnJJ9H4Za5ENNYKsnI2U6D9tqlJnlOrvknO4OxfaI1Zi5Ww8nrXUFj5fR07dkRWVhYqKipw7733omdPHeo8WI0WGRBK6jtpRetMIC1G/wdbl6Kvb1QWghYZQEWC245oO5e+gtPy9+2irIyE2owZX2rqE+0RnFQsGtkTQmf6ndNN3Wtn6nz5Lhh/25MetYHU8LfffrsFWLMZqKmTg+uaOvnvb7eEvzzRz697dvjLCrYf61FHqYuKUhsaEA5mTp06hcmTJyMzMxO5ubn4y1/+AqfTiUcffRTdu3fHunXr8Morr+jZV2vQIgMinPpOWtAiE0iLszvB1qXo6xuVhaBFBpBoIValBVuzBFOzdx4Qr18EAG3aKOtHIGrrE8WqpubQmX57VM6qbGaShr9sJj1qA6nhu99+uyVw+ZL6xvADmo4dxdrla1C41t96d9GjjlKByiKoYRIOZh555BF8/fXXmDJlCjp06ID7778fl19+OVavXo1PP/0UpaWluO666/TsqzWozSLwpEW9jXCFmwmkRabLgSADV0WzHIzKhhCtRRWsnWg9HaV1d/RaB1rVzdEjYyLaBctAaWoKXScskKNH1T1PC/6y9IzOIgrEM8NUZP3WN4ZXn050n92+Xf0yXILtf2pqfIViUi024WDmk08+wbJly7Bo0SJ8+OGHkCQJvXr1wpdffonRo0fr2UdrUZtF4EmLehtaCCcTSItMl/IgKcWiWQ5GZUOI1qIK1k6vLCK91oFWdXP0yJiIdsG2gXDqov14WP1zw+UvS8/oLKJAPDNMtdjXQxHdZ09oEGwE2/+0ylj0ZFItNuFg5siRI+55ZLp3746kpCTcdtttunWMDBBOJlBsDJD/hWgtqmDtROvpRGLdHdJWsAwULeqimcFflp4Vs4i02NcjRRQda4QHHDidTtg9RnHHx8ejjVbX1Mkc4WQC2RBbAU2SXSzDxrdmVX29/OtTyUE7EuvukLaCZaCIbotW4y9LTzRL0Ehq9/VIFB8XNQGN8FFTkiRMnToVV111Fa666irU19fjzjvvdP/tuimxYMEC/OpXv0JaWhqysrIwadIk/PCD96ltSZIwd+5c5ObmIjk5GWPGjMF2La4jqqU2i8CKwskE0iLTpSDIwFXRLAejsiGSBQcAe9asWrUB+Hab8oO10nWr1zrQKqNTj4yJaBdsGwinLlovE6fO8Jelp0fdODU8M0xF1284n4PoPttRZdkKT8H2P60yFj2ZVItNOJiZMmUKsrKykJ6ejvT0dNxwww3Izc11/+26KVFSUoJp06Zh3bp1WLlyJZqbmzF+/HivGk8LFy7EM888gyVLlqC0tBQ5OTkYN24carUanKiU6Ch0q0tKDC8TSIuzcl2CpBSLZjkYkQ2x5yBw7KfQ7eLjfpmDYtUG9ZNVKa27o9c6kCC/j3DpkTER7YJloKid5yQ+DujUSd1zteAvS0+PunFqeGaYiq7fcOabEd1nzzlH/TJcgu1/RzRIaPFlUi024W+zZcuWab7wzz77rNUysrKysGHDBlxwwQWQJAmLFy/GnDlz3Gd9Xn31VWRnZ+PNN9/EHXfcoXmfQoqWzAxJkm+hUkAD0SKDprIycFpxsGwO33bBDvzhcjqBQ4JTirc45faNjeHNunnmjLJgMVQdn3A4JflSWThfOrFQgFZr338fuCjkRpXV1Vuc5n4WFRWtCyyK7ud6q67+5Ut/3z6x5+zbB3Trpm55olllWlyFOHHC/49wJcc2JdavB4qNLzZpqYvz1T8HCh1+ripdXl6OiooKjB8/3t3G4XBg9OjRWLNmjd/XaGhoQE1NjddNU9GSmdHQFLr+SzBaZNAEqylkdP2gQI4IFnH0bB9uhobS9xTO5yjC6PdDwLEgRVhrwxjjYOZn8YOfiRCtsm14Htf3C56tEG3nj2hWmRbZTIEycJUe20SZFC9bJpiRJAkzZ87E+eefj/79+wMAKioqAADZ2d6zIGZnZ7sf87VgwQKvy175Wkw6FK2sPIhQ9MyG3nVHziqsYH22IfwBjUrfkt6fo9Hvh/Rjtc8iRuoGWZLSY5vFWSaYueeee7Blyxa89dZbrR6z+VwKkSSp1X0us2fPRnV1tft2UIsS6tHKyvVfROuJ6F13JFnhbLzJjuD1mUQofUt6f45Gvx/Sj9U+ixipG2RJSo9tFmeJYObee+/Fhx9+iK+++gp5eXnu+3N+vr7qexamsrKy1dkaF4fDgbZt23rdNBUtmRkOe+j6L8FokUETrKaQ6Ch7PUbje8oVLBXg2T7cDA2l7ymcz1GE0e+HgOwgpT7Swjhsm/lZ9M5rfZ9Vtg3P43pXwSQP0Xb+iGaVaZHNFCgDV+mxTZRJM7aYGsxIkoR77rkH7733Hr788ksU+NR0KCgoQE5ODlauXOm+r7GxESUlJRg5cqTR3ZVFS2ZGjy7qB/8C2mTQBKspJDqoV8/BvwAQFwfkCRZ7y8uW2yclhfeLU2mmmM0G9OuhfnnBxNnCzzjhfFTKBRr8CwBnVF6ayUiXPwuzzob4Dv4F9N9/RXke10UH9aod/AuIZ5Vpkc0UKANXybFNCRMG/wImBzPTpk3D66+/jjfffBNpaWmoqKhARUUFzv5cL8Jms2HGjBmYP38+3n//fWzbtg1Tp05FSkoKrr/+ejO7HvmUFBz0R0n1ZbUyQgSOoR7XyrEqsXanPAbipmjwi0qJYzqkWAIc02A14aT81/w8MlPP7LdAjNpX1bAbVKzWivTIZjKJqcHM0qVLUV1djTFjxqBTp07u2zvvvONuM2vWLMyYMQN33303iouLcfjwYaxYsQJpaTqfWg/EzEJtWlObZi5J4QdDQPCinS0tQFWI/lVVy+301NgoVzEW4Sre2dws/1+t00EyWfwRWVfhqAwz6+GkAYFvtPE31q++PrzgsqlZ3ufNiE8D7avhbltaaGqW93OXYAVwPYm280c0Jb28XP0yXALNyfaTwNxZamzbps/rhmD6ZSZ/t6lTp7rb2Gw2zJ07F0ePHkV9fT1KSkrc2U6mMLNQm9bUpplrlQocrGjnXsGB26Lt1NqssNji93vDK+AJABv8FOQLRu91ECyFXsQWDQLfWLPXzy9mLYoymjm1hL/tNNxtSyue+3mwArieRNv5I5qSfkDwrHAwgQrGbg3zOBVIlTmV0GP4/BqpZkRKd51g2qBoO7UaBM/KuIRTvFMtvdcBWYPVahgpZeXtVOl+Hi5evtUcgxlSTutU4LNn5V8qLU55yvWhfYEUh/cYlEBSdE4vdCQAdQq+RBrrgfS2QF2Yv04OVcjZBnECJ09F1xVFNisWZVRC7301HA6NvwobG+WzPQ3N8msP6g0kehT2jbMxoNGYJVKzI4qZhdq0pjbNXKtU4HO6AV+vB77b/kvl1han/LdozZDuOk+KOKi3svbNAKo0mHV6zyFg1Ua5LlQoeq+DYCn0IgbqlGkVzbr7yTLRoiijmVNL+NtOw922tOK5nwcrgOspULs1ZcDaLfKZqJYW+d+1W+T7XYxMSR8S4Bg2IIxCw8FkmFNvi8GMUmYWatOa2jTzcFK6Pe3YF/5gxPgwJ3QLxfPXlBkOHQsd0Oi9DoKl0Ito316bfsQSfzOXa1GU0cypJfxtp+FuW1qwJ3jv58EK4Hry125NWeCEgabmXwIaI1PSAyXL/Fw2SHMmjWllMKNUg4Wv+yp15Ii652lRZBLQJqtC7+rpVRoMwAvXoWNyUbhA9M4WCidrAwC0ro8WC0rWt75Pi23RzEKT/o6dZmeHJsQDIwd73yea5ePbTiTz0ZU5FSyTU2uBsiP1ymbyt+0agMGMUpsUZppY2S6VwYwWRSa1Emikvla2aZAaqYVgReH0zhYKJ2sDADb9qE0/Yp0W26KZhR39HTvNzg719yNBNMvHt51o5uPmH4JncmotUHakXtlMJmEwo1SjwaPeiYCoKwpHJjFzzKkVj51aDsIVzYgyOnMqRjCbSanEBKDBwtWmKTpFWVE4MokN5gU0iQnArl3AER0neFRKy9IOopmPSjMkSQjPzChVFKRmSqQpzFX3PC2KTGol0Eh9rfQvCN3GCMGKwumdLSSa3RFIUS9t+hHrtNgWzSzs2NBkrUAGAJL8/EgQzfLxbSea+ag0QzJcQwN8Z+mVzWQSBjNKOaLoF3KuymBGiyKTWtF7zExGhr6vL8KG4PPN6J0tJJrdEYjWletjwWg/xfq02BbbtJG3J5LV1QMbdnjfJ5rl49tONPNx7RaxdlpJDVCBXa9sJn/brgEYzCildy0gI+1ROXC02WLXfPXMlqk3Z2puLxKCZ9HpnS2kNuvNZd8+TboRU/xlhOzSoBRBba2542asyFVTzWW3YIKDbzsrHCv8CdQvvb7LmM0UIfSug2OkQypTesOtPaQ1PbNltKiHo4VgWXR6ZwupzXpz2W9gGmo00+ISjd5nMiOV5zHt8Cmx5/i2s8qxwlegfkXTdxkYzChn5foiRjGj/pBZrDJ9vBUzQYiihRbHNKscK3wF6leUfZcxm0kp1sEBkhPDrz1kJSdOeM/7cE43oGNH+f9xccEnrDNKIndVIt0kazDTt1VrZ9kDzBAeZd9lPDOjlN51cIyUp3LgaB+LjYIPJ1umZH3rCay275PvL1lvjUAGCJ5Fp3e2kNqsN5euHbXpR6zLNbEUQbTzPKZ1bif2HN92WtTO0kOgfkXTdxkYzCindx0cI/VQmdKbkKDt/AzhUpstY9JANVWCZdHpnS2kNuvNpVs3TboRU/xlhBSaWCQymtkT5GOaS0/BqSd822lRO0trNgTuV3y8PpltzGaKENUWmychHGprvTQ3W6t8vZrPxMjaKFoIVoNK7yyKtWEGfeFmQ8Uif4E2s8L00dTsfQZWdHv3bddkwclU4+IAKcCxurFRn8w2ZjNFiDIN0iOtQm2tF6tlM6n5TIysjaKFYFkoemdRhDs2MtxsKJIxK0w/nrXPRLd333bbLPjd0OIEqgP8EBKtJRUhGMyQcrGUzRQJrDjokCiSaFH7rN6CZ2YAoDFAv6KsRhSDGVJOi5H/pJ1A2QpEJEaL2mdJ9vBfQw+JAfrliK4MSQYzSg2OokF4Smu9tLQA234AqnSecVapwYXAqVO/ZCCVrJf/DuacbgZ0TEPBalDpnUURbuwabjYUyZgVph/P2mei27tvu/4W/G6IjwPS0/w/ZnSNKJ0xmFEqPYrSI5XUetm2C1i9Caiy4LwEZbuAzT5Ti2/eHXwgWscI+2JIC3BAAvTPohgRZnZCuNlQschfRgizwvSRl+1d+0x0e/dtZ7cDSRY7a927ALAFSFlKTJQzubTGbKYIEUnpvKGIvpdtu4CqCM3iiqbPK5BvDS5cRxRNDh1T9zx/x5Z6i40n3LU/+ONNOoyb+ZrZTNZ3+rTZPdDemTPBH29pidxAxsXfJadQ79tqAgVlTU36H0DDnY7g+HFt+hFL/H3e62MgMDeL5zFCyQ8gz+dZcQqCpmY5BdsftYWGQ5EAnD2rz2sHwWBGiQ1Biv1Fqg0h0nqjoRiZ7yUoIPT7jhRGpIOGOx3BjhC/DklMhMXfEcXfMULp86w6BUGgFGy1hYZFmHB8ZTAT60JNmhRlxcjcLDTnX1ismg5KRNZgRgp2i/FlYBjMxLpQ01mnaJCyaDWVlaHbRILGRmvOOkpE1mFGCna88aEFgxklhgYp9hephoZI642GYmSDPGqolKwHdh4wry9aWVMGrN1izBmmcKYjiIUB2EZpY3YHopjrGKF0e/U8tlh1CgJ/Kdh675ehvld0wGBGidRUs3ugvTYhjpDRUFizXTv530j+YvVMd1xTpk8WQiBqpyOI5PVtNn/prcXmpLzGhHbt1G2vrmMLYN0pCBJ90sWN2C+Tk/Vfhg8GM0ocU5nCZ2XOENc2ayw2QZ4a+/dH/qWlsjL538ZGYwMZQF0WDYsihsffFw6DQ/2ozUT63iMppK5Om75ozTObyag+mvC9wWBGie+jILPH15EQX/KbfjSmH3radzzyLy1V/xzAmFEcTk0WDYsiUiRRm4l0zGO6DqtmSHoeM4zqownfGwxmYp0WBdbIOFFWHI4oajgtmiLpecywah81EF2Vpkg5fwXWjh2LzrNQ0cAeJ09kSETWEmezZrDgmc1k1T5qgGdmlOgTBZk9vjwLrAHydfloC2S6ZQJ9u5jdi/CkJwCrNpgzr4yaLBoWRaRYkO2RFGJCBo8Qz2wmo/pY1MuY5XhgMKNEdrbZPdCeZ4G1aB1g2LUrkJUVup2V1baY94tKTRYNiyKGx182k0kF/CiIPh7TdaSkmNePYDyzmYzqI8fMWJxetSzMVFUl/xuNmVouBw5Efn0gM08NqwlydxlQZiGaMZspMqz2+Exqa83rRzCuYzwANBs45s7gWlUMZpTQs5aFWbaVy/9G26UlT+WVrA9ktCMRXpyUSITn8LWNJmQainAd4wHg+73GLdfgWlUMZoiIiGLB2QAVtKMAgxkiIqJYkJwYuk2EYjCjRF57s3ugvf4F8r/RmKnlUpAF9Otqdi9iS67KEghEkcSz2ssQPzWQrMB1jAeAPt2NW67BtaoYzCjRo4fZPdBeRob8bzRmarl06QJkZprdi8ilJoumMIzilMRspkhxvsdnkpZmXj+CcR3jASDBwKnlDK5VxWBGiTVRmE1w+LD3v9Ho2DGgmgNSVVOTRWNwJkPUYTZTZPD8TKyaEerZRyNrpp1RUwdFPQYzSpgwX5nudh/1/jcafX8QKGOqsKEMzmQgMl0kZIQaWTPN4FpVDGaIiIhIWwZPjcVghoiIiLRlM3ZxDGaUsJvdAR307OT9bzTqkw8M5oBUQxmcyUBkukjICDWyZprBtaoYzCgxMgqzCTp3lv8tt+jgNS1kZwPpTBVWTU0WjcGZDFGHmUuRwfNzsmpGqGcfjayZ1kZNhVr1GMwoEY3ZBKdOAQ0NQIvT7J7op2Q9cDIKS1EYRc12f+CA9v2IJb7rfMcOc/pBwXl+TuXlgduZybOPDQ3GLdfg7C4Dk87JkjbvBhzReP3Mx5YoLBJqZeWVZvcguhyvM7sHFMqBqtBtzLbpe+OW9f1BQ89W8cwMAY0GVlIlIiJzRPGxnsEMAYk8QUdEFPWi+FhvajDz9ddfY+LEicjNzYXNZsMHH3zg9bgkSZg7dy5yc3ORnJyMMWPGYPv27eZ0NlpltAUGxkCmz8AoLEVhZUZmTcSCzBSze0ChdMkI3cZsRX2MW5bB2V2mBjNnzpzBoEGDsGTJEr+PL1y4EM888wyWLFmC0tJS5OTkYNy4caitrTW4pz+LxgyDqhqgNMoHF44uBtpHYZFQoyjd7vccNHam0Wjku8779TOnHxSc5+dUUBC4nZk8++hwAPEGfe0bnN1l6jmnCRMmYMKECX4fkyQJixcvxpw5c3DVVVcBAF599VVkZ2fjzTffxB133GFkV4lIxJ6DwKEoTvMninRRmrlq2TEz5eXlqKiowPjx4933ORwOjB49GmvWrDGnU9GYmh0LStbzswuH6LpzOhnIaMV3nf/0kzn9oOB2edR8s2oxW89tqabGuOUauSxYOJipqKgAAGT7nKrKzs52P+ZPQ0MDampqvG5EZIAjTMfWzda9ZveA/DniEcBEQjHbTT9G57Jg4WDGxWbzLvAgSVKr+zwtWLAA6enp7lt+fgRMMU0UDc4aOCEXEZEHywYzOTk5ANDqLExlZWWrszWeZs+ejerqavft4MEIKMtOFA2SHWb3gIhilGWDmYKCAuTk5GDlypXu+xobG1FSUoKRI0cGfJ7D4UDbtm29bkRkgNwss3sQvQZ0N7sH5E+uR823SChmW9QrOpcFk7OZTp8+jd27d7v/Li8vR1lZGTp06IAuXbpgxowZmD9/PgoLC1FYWIj58+cjJSUF119/vTkdHl3MgaSRyJWayM9OHdHU7DjL/jaKPL7rvEMHABw3YzmFHgGMVYvZem5LRv64N/hEgqlHn/Xr16OoqAhFRUUAgJkzZ6KoqAiPPvooAGDWrFmYMWMG7r77bhQXF+Pw4cNYsWIF0tLSzOkwvwwjU8l6YNs2s3sRuUS3e+4f2vFdl42N5vSDgvP4MY6mJvP6EYzntmTkPnrC2LmmTD0zM2bMGEiSFPBxm82GuXPnYu7cucZ1iqJTVb3ZPSBSb/MPZveA/Dl8Cuj58/+3RUA2k5G27wNGGzcTOM8LExFZXUP0FgiMGvUWPTMTIxjMEBFZnSN6CwRGjSS72T2IaQxmiIisro5z+FiWaxxK/wjIZjLSOd0MXRyDGSWisdAkUSii2z33D+14rksOrLa+kvWA3aJnZjy3JSP30Y7GjZcBGMwow4MKxSLR7b6eg6w141rnZ86Y2w8St2+f2T3wz6xspm+N/b5kMENE2ti40+weRJ8NXKcRY7+xqciWZ/BvGwYzRKSNphazexB9As9cQUQeGMwQkTbs8Wb3IPoErqlLpMxPP5ndA10xmCEibQzpa3YPos9QrtOI0dXYAa+KlKwHthpcDiPJ2MUxmFGC2RoUi0S3+ySDj17RzLXO27Qxtx8krls3s3tgLcOM/b5kMKMEs5koFolu92u4f2jGtc55zIkc/Ky8Gbw+GMwQkTY4mzsRmYTBDBEREUU0BjNEREQU0RjMEJE2LDqbOxFFPwYzSjCbiWKR6HY/kvuHZlzrnMecyMHPypvB64PBDBEREUU0BjNKMPWOYpHods/9QztMzY48/Ky8MTWbiIiISByDGSIiIopoDGaIiIgoojGYISIioojGYEYJpt5RLBLd7rl/aIep2RTpmJptYRytTrFIdLvfvVvffsQSZjNRpGM2ExFFpMOnzO4BEcUoBjNEREQU0RjMEBERUURjMENE2ujczuweEFGMYjCjBDMLKBaJbvc9e+rbj1jCbCaKdMxmsjBmFlAsYm0m4zGbiSIds5mIiIiIxDGYISIioojGYIaIiIgiGoMZIgptz0Gze0BEFBCDGSWYWUCx6tCx0AEN9w/tMJuJIh2zmSyMmQUUyw4dA5zOwI9z/9AOs5ko0jGbiYgs60il2T0gImqFwQwRiTvbYHYPiIhaYTBDROKSHWb3gIioFQYzRCQuN8vsHhARtcJgRglmFlCsiwtyyOD+oR1mM1GkYzaThTGzgGJdsH2A+4d2mM1EkY7ZTERERETiGMwQERFRRGMwQ0RERBEtwewOEFGE4TgOY3A9EwnjmRklmFlAREQUGrOZiIiIiMQxmFGCp32JiIhCY2o2ERERkTgGM0RERBTRGMwQERFRRIuIYOa5555DQUEBkpKSMHToUKxatcrsLhEREZFFWD6YeeeddzBjxgzMmTMHmzZtwqhRozBhwgQcOHDA+M4wNZuIiCg0pmZ7e+aZZ3DrrbfitttuQ9++fbF48WLk5+dj6dKlxnfm4EHjl0lERBRJEu2AJBm6SEsHM42NjdiwYQPGjx/vdf/48eOxZs0av89paGhATU2N100ze49p91pERETRqLEJqK41dJGWLmdw4sQJtLS0IDs72+v+7OxsVFRU+H3OggULMG/ePCO6Fx0CnQrctR84clzfZfctALIyxNuL9ik3EyjsGv7rkLdEOzBikFhbzskUHte+wW01svge06ywH8THAy0txi+3scnQxVn6zIyLzWbz+luSpFb3ucyePRvV1dXu20FeGlIn2aH/MhLtytqL9ilUOyPeWzRKUvB52eP160cscO0b3FYji9JjmhEcJp2zMHhdWDqY6dixI+Lj41udhamsrGx1tsbF4XCgbdu2XjfNdPe/zIjVs1Pgx3Kz9F22ww6kpyl7jmifQrXT+71Fq/6F4m2H9NWvH9Eu0WPf4LYaOfwd04p6mdMXT4N6G7/MRBXH9zBZOphJTEzE0KFDsXLlSq/7V65ciZEjRxrfofx845epp86dAz8WFwfk6Ri89egCBDi7FpBIn/Ky5Xbhvg55S0oE7Ap+aSUlAXEKP1+S9fTYN7itRg5/xzQtf0yrkZoCJCYavw31VHF8D5OlgxkAmDlzJl566SW88sor2LlzJ+6//34cOHAAd955pzkdipb0bJH30SNf+U4QZwOCbcMOO9CvB5DZXtnrivQpL1t+PNzXiY+T+9ivR+jAKBSHhqdawzk22BPkzzw1RflzkxKBYQOVP2/UUAY0SsTZ/O8bavbDQDLS5e2An4t2Qh3TzPrOSE0BhvaT/6/lNhRKOMf3MNgkyeD8KRWee+45LFy4EEePHkX//v3x3//937jggguEnltTU4P09HRUV1dre8np4MHg2U32eKC5BfBcu50zge758hdkfT3w7Tbly81uD6SkAA2N8peMIxE4cFReVlKi/Iu4vgFIiAPSU4G9R7yf37NT8DMy/jidwJFKYO8h7/fTIweoawJq6uQvy7xsoEO6HJHX1wMbdwJNLXJw0C1X3rnS07SJ2F19OtsgjyvIzVIXeDidwOFjQPVpeaBcdgbQvu0vfZQk4GQ1sO8QUFsv32ezAbkZwOET3q/Vvg1wTi+g9ow8+M11qtVmA06dAjbv/qVtigNodsrjUPoXysv+cS9w7JT3azrsQPt0oGe+3KaxEdj8A9DQLF8LP6cH8EM5UHPW//tLcgBFveVfZy7NzcD3e4GTNYDTo22cDWiXBjQ2y/139U3JGRl/6uuBDTvk9+vLHi9vI1qwwXv7dMlIAxxJ8j7xUy1wuk78NeNsgDPAITLh533c977MdkDFT61TU7PaAwkJP++nDnm7a2iS/++57wTiu83ndJRfo/KkvKz0NKB7Z+DocaCqWh70aYuT+5TikI898R5jmerrgdJt3tuA1voXAEergLONQHIikJ8NlO365XGHXf7MkuxAQS6wZbf/z9CfZAeQ1UHe78/U/fy52uTl9Okur+uaGmDTj/6f3zYFGNhbXucHj8rrsaVF3t7TUoCUJPlLecceoN5jf0hIkDN1fPfxUHz70jEdaJMM7PdJZhnQHejQAWhqArbtktddHOR+1dX/sl3Fx/+yLdWelvdbm03uT7+f378v1zZUVy/vL6ltgMpK4FSA44eoOBvQuxuQ2UHTMzJKvr8jIpgJh27BDBEREelGyfe35S8zEREREQXDYIaIiIgiGoMZIiIiimgMZoiIiCiiMZghIiKiiMZghoiIiCIagxkiIiKKaAxmiIiIKKIxmCEiIqKIZlJtcOO4JjiuqakxuSdEREQkyvW9LVKoIOqDmdraWgBAfrRVvCYiIooBtbW1SE9PD9om6mszOZ1OHDlyBGlpabBpXJK8pqYG+fn5OHjwIOs+eeB68Y/rxT+uF/+4XvzjevEvGteLJEmora1Fbm4u4kIUEo76MzNxcXHIy8vTdRlt27aNmo1HS1wv/nG9+Mf14h/Xi39cL/5F23oJdUbGhQOAiYiIKKIxmCEiIqKIxmAmDA6HA4899hgcDofZXbEUrhf/uF7843rxj+vFP64X/2J9vUT9AGAiIiKKbjwzQ0RERBGNwQwRERFFNAYzREREFNEYzBAREVFEYzCj0nPPPYeCggIkJSVh6NChWLVqldldMt3XX3+NiRMnIjc3FzabDR988IHZXTLdggUL8Ktf/QppaWnIysrCpEmT8MMPP5jdLdMtXboUAwcOdE/wNWLECHz66admd8tyFixYAJvNhhkzZpjdFdPNnTsXNpvN65aTk2N2tyzh8OHDuOGGG5CRkYGUlBQMHjwYGzZsMLtbhmIwo8I777yDGTNmYM6cOdi0aRNGjRqFCRMm4MCBA2Z3zVRnzpzBoEGDsGTJErO7YhklJSWYNm0a1q1bh5UrV6K5uRnjx4/HmTNnzO6aqfLy8vDUU09h/fr1WL9+PX7961/jiiuuwPbt283ummWUlpbihRdewMCBA83uimWcc845OHr0qPu2detWs7tkupMnT+K8886D3W7Hp59+ih07duDPf/4z2rVrZ3bXDMXUbBWGDRuGIUOGYOnSpe77+vbti0mTJmHBggUm9sw6bDYb3n//fUyaNMnsrljK8ePHkZWVhZKSElxwwQVmd8dSOnTogP/6r//CrbfeanZXTHf69GkMGTIEzz33HJ544gkMHjwYixcvNrtbppo7dy4++OADlJWVmd0VS3n44YfxzTffxPzVAZ6ZUaixsREbNmzA+PHjve4fP3481qxZY1KvKFJUV1cDkL+4SdbS0oK3334bZ86cwYgRI8zujiVMmzYNl112GS666CKzu2Ipu3btQm5uLgoKCvD73/8ee/fuNbtLpvvwww9RXFyMa665BllZWSgqKsKLL75odrcMx2BGoRMnTqClpQXZ2dle92dnZ6OiosKkXlEkkCQJM2fOxPnnn4/+/fub3R3Tbd26FampqXA4HLjzzjvx/vvvo1+/fmZ3y3Rvv/02Nm7cyLO8PoYNG4bXXnsNn3/+OV588UVUVFRg5MiRqKqqMrtrptq7dy+WLl2KwsJCfP7557jzzjsxffp0vPbaa2Z3zVBRXzVbLzabzetvSZJa3Ufk6Z577sGWLVuwevVqs7tiCb1790ZZWRlOnTqFd999F1OmTEFJSUlMBzQHDx7EfffdhxUrViApKcns7ljKhAkT3P8fMGAARowYgR49euDVV1/FzJkzTeyZuZxOJ4qLizF//nwAQFFREbZv346lS5fipptuMrl3xuGZGYU6duyI+Pj4VmdhKisrW52tIXK599578eGHH+Krr75CXl6e2d2xhMTERPTs2RPFxcVYsGABBg0ahGeffdbsbplqw4YNqKysxNChQ5GQkICEhASUlJTgL3/5CxISEtDS0mJ2Fy2jTZs2GDBgAHbt2mV2V0zVqVOnVj8A+vbtG3MJKQxmFEpMTMTQoUOxcuVKr/tXrlyJkSNHmtQrsipJknDPPffgvffew5dffomCggKzu2RZkiShoaHB7G6YauzYsdi6dSvKysrct+LiYkyePBllZWWIj483u4uW0dDQgJ07d6JTp05md8VU5513XqvpHn788Ud07drVpB6Zg5eZVJg5cyZuvPFGFBcXY8SIEXjhhRdw4MAB3HnnnWZ3zVSnT5/G7t273X+Xl5ejrKwMHTp0QJcuXUzsmXmmTZuGN998E//7v/+LtLQ09xm99PR0JCcnm9w78zzyyCOYMGEC8vPzUVtbi7fffhv//ve/8dlnn5ndNVOlpaW1Gk/Vpk0bZGRkxPw4qwcffBATJ05Ely5dUFlZiSeeeAI1NTWYMmWK2V0z1f3334+RI0di/vz5uPbaa/Hdd9/hhRdewAsvvGB214wlkSp/+9vfpK5du0qJiYnSkCFDpJKSErO7ZLqvvvpKAtDqNmXKFLO7Zhp/6wOAtGzZMrO7ZqpbbrnFvf9kZmZKY8eOlVasWGF2tyxp9OjR0n333Wd2N0z3u9/9TurUqZNkt9ul3Nxc6aqrrpK2b99udrcs4aOPPpL69+8vORwOqU+fPtILL7xgdpcMx3lmiIiIKKJxzAwRERFFNAYzREREFNEYzBAREVFEYzBDREREEY3BDBEREUU0BjNEREQU0RjMEBERUURjMENEEWXfvn2w2WwoKysL2m7MmDGYMWOGIX0iikVff/01Jk6ciNzcXNhsNnzwwQeKnj937lzYbLZWtzZt2ijuC4MZItLF1KlT3Qcnu92O7t2748EHH8SZM2fCet38/HwcPXrUPb3/v//9b9hsNpw6dcqr3XvvvYfHH388rGURUWBnzpzBoEGDsGTJElXPf/DBB3H06FGvW79+/XDNNdcofi3WZiIi3VxyySVYtmwZmpqasGrVKtx22204c+YMli5dqvo14+PjkZOTE7Jdhw4dVC+DiEKbMGECJkyYEPDxxsZG/Od//ifeeOMNnDp1Cv3798fTTz+NMWPGAABSU1ORmprqbr9582bs2LEDzz//vOK+8MwMEenG4XAgJycH+fn5uP766zF58mR88MEHaGhowPTp05GVlYWkpCScf/75KC0tdT/v5MmTmDx5MjIzM5GcnIzCwkIsW7YMgPdlpn379uHCCy8EALRv3x42mw1Tp04F0Poy08mTJ3HTTTehffv2SElJwYQJE7Br1y7348uXL0e7du3w+eefo2/fvkhNTcUll1yCo0eP6r+iiKLQzTffjG+++QZvv/02tmzZgmuuuQaXXHKJ137n6aWXXkKvXr0watQoxctiMENEhklOTkZTUxNmzZqFd999F6+++io2btyInj174uKLL8ZPP/0EAPjjH/+IHTt24NNPP8XOnTuxdOlSdOzYsdXr5efn49133wUA/PDDDzh69CieffZZv8ueOnUq1q9fjw8//BBr166FJEm49NJL0dTU5G5TV1eHRYsW4e9//zu+/vprHDhwAA8++KAOa4Iouu3ZswdvvfUW/vGPf2DUqFHo0aMHHnzwQZx//vnuHyaeGhoa8MYbb+DWW29VtTxeZiIiQ3z33Xd48803ceGFF2Lp0qVYvny5+xT1iy++iJUrV+Lll1/GQw89hAMHDqCoqAjFxcUAgG7duvl9zfj4ePflpKysLLRr185vu127duHDDz/EN998g5EjRwIA3njjDeTn5+ODDz5wX6NvamrC888/jx49egAA7rnnHvzpT3/SahUQxYyNGzdCkiT06tXL6/6GhgZkZGS0av/ee++htrYWN910k6rlMZghIt18/PHHSE1NRXNzM5qamnDFFVfg3nvvxT//+U+cd9557nZ2ux3nnnsudu7cCQC46667cPXVV2Pjxo0YP348Jk2a5A5C1Ni5cycSEhIwbNgw930ZGRno3bu3e5kAkJKS4g5kAKBTp06orKxUvVyiWOV0OhEfH48NGzYgPj7e6zHPcTIuL730Ei6//HKh8XD+MJghIt24zsLY7Xbk5ubCbrdj8+bNAACbzebVVpIk930TJkzA/v378cknn+D//u//MHbsWEybNg2LFi1S1Q9JkgLe79kPu93u9bjNZgv4XCIKrKioCC0tLaisrAw5Bqa8vBxfffUVPvzwQ9XL45gZItJNmzZt0LNnT3Tt2tUdKPTs2ROJiYlYvXq1u11TUxPWr1+Pvn37uu/LzMzE1KlT8frrr2Px4sV44YUX/C4jMTERANDS0hKwH/369UNzczO+/fZb931VVVX48ccfvZZJROJOnz6NsrIy95xP5eXlKCsrw4EDB9CrVy9MnjwZN910E9577z2Ul5ejtLQUTz/9NP71r395vc4rr7yCTp06Bc2MCoVnZojIUG3atMFdd92Fhx56CB06dECXLl2wcOFC1NXVuQf/Pfrooxg6dCjOOeccNDQ04OOPPw4YdHTt2hU2mw0ff/wxLr30UiQnJ7c6jV1YWIgrrrgCt99+O/7f//t/SEtLw8MPP4zOnTvjiiuu0P09E0Wj9evXu7MJAWDmzJkAgClTpmD58uVYtmwZnnjiCTzwwAM4fPgwMjIyMGLECFx66aXu5zidTixfvhxTp05tdTlKCQYzRGS4p556Ck6nEzfeeCNqa2tRXFyMzz//HO3btwcgn22ZPXs29u3bh+TkZIwaNQpvv/2239fq3Lkz5s2bh4cffhg333wzbrrpJixfvrxVu2XLluG+++7D5ZdfjsbGRlxwwQX417/+1erSEhGJGTNmTNDLsHa7HfPmzcO8efMCtomLi8PBgwfD7otN4gVhIiIiimAcM0NEREQRjcEMERERRTQGM0RERBTRGMwQERFRRGMwQ0RERBGNwQwRERFFNAYzREREFNEYzBAREVFEYzBDREREEY3BDBEREUU0BjNEREQU0RjMEBERUUT7/8ToxHYuBsvFAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plot your graph and show it here (matplotlib/seaborn)\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#Loading CSV\n",
    "data=pd.read_csv('~/biol6150/ProjectSubmissions/Group8/Project3/chr20_reads.csv', header=None, names=['position', 'read count'])\n",
    "print (data)\n",
    "\n",
    "#Create line plot (position vs read count)\n",
    "plt.plot(data['position'], data['read count'],marker='o', linestyle='-', color='pink')\n",
    "\n",
    "#graph labels\n",
    "plt.xlabel('Position')\n",
    "plt.ylabel('Read Count')\n",
    "plt.title('Position vs Read Count')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8b119fe",
   "metadata": {},
   "source": [
    "# **2. Variant calling**\n",
    "### *Total Questions: 3*\n",
    "### *Total Points: 10+5+2+10 = 27*\n",
    "\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68860d7d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>2.1 Call the variants (SNPs and Short Indels)</h3>\n",
    "    <p><b>2.1.1</b> Write the command used and justify the flags you have used using one sentence.</p>\n",
    "    <p><b>2.1.2</b> Informed decision.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e6bf56ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VarScan v2.4.6\r\n",
      "\r\n",
      "***NON-COMMERCIAL VERSION***\r\n",
      "\r\n",
      "USAGE: java -jar VarScan.jar [COMMAND] [OPTIONS] \r\n",
      "\r\n",
      "COMMANDS:\r\n",
      "\tpileup2snp\t\tIdentify SNPs from a pileup file\r\n",
      "\tpileup2indel\t\tIdentify indels a pileup file\r\n",
      "\tpileup2cns\t\tCall consensus and variants from a pileup file\r\n",
      "\tmpileup2snp\t\tIdentify SNPs from an mpileup file\r\n",
      "\tmpileup2indel\t\tIdentify indels an mpileup file\r\n",
      "\tmpileup2cns\t\tCall consensus and variants from an mpileup file\r\n",
      "\r\n",
      "\tsomatic\t\t\tCall germline/somatic variants from tumor-normal pileups\r\n",
      "\tmpileup2somatic\t\tCall germline/somatic variants in multi-tumor-normal mpileup (beta feature in v2.4.5)\r\n",
      "\tcopynumber\t\tDetermine relative tumor copy number from tumor-normal pileups\r\n",
      "\treadcounts\t\tObtain read counts for a list of variants from a pileup file\r\n",
      "\r\n",
      "\tfilter\t\t\tFilter SNPs by coverage, frequency, p-value, etc.\r\n",
      "\tsomaticFilter\t\tFilter somatic variants for clusters/indels\r\n",
      "\tfpfilter\t\tApply the false-positive filter\r\n",
      "\r\n",
      "\tprocessSomatic\t\tIsolate Germline/LOH/Somatic calls from output\r\n",
      "\tcopyCaller\t\tGC-adjust and process copy number changes from VarScan copynumber output\r\n",
      "\tcompare\t\t\tCompare two lists of positions/variants\r\n",
      "\tlimit\t\t\tRestrict pileup/snps/indels to ROI positions\r\n",
      "\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "#Use VarScan for calling variants. Look up the appropriate command for this.\n",
    "\n",
    "#Running VarScan\n",
    "!varscan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee12252e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: No p-value threshold provided, so p-values will not be calculated\n",
      "Min coverage:\t8\n",
      "Min reads2:\t10\n",
      "Min var freq:\t0.2\n",
      "Min avg qual:\t15\n",
      "P-value thresh:\t0.01\n",
      "Reading input from /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup\n",
      "70626195 bases in pileup file\n",
      "1484 variant positions (1469 SNP, 15 indel)\n",
      "32 were failed by the strand-filter\n",
      "1484 variant positions reported (1469 SNP, 15 indel)\n"
     ]
    }
   ],
   "source": [
    "# Using mpileup2cns because this command will identify variants of the mpileup file we created in section 1 that summarized bases of aling reads\n",
    "!varscan mpileup2cns ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup --min-reads2 10 --output-vcf > ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d5e4e326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: No p-value threshold provided, so p-values will not be calculated\n",
      "Min coverage:\t8\n",
      "Min reads2:\t10\n",
      "Min var freq:\t0.2\n",
      "Min avg qual:\t15\n",
      "P-value thresh:\t0.01\n",
      "Reading input from /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup\n",
      "70626195 bases in pileup file\n",
      "1484 variant positions (1469 SNP, 15 indel)\n",
      "32 were failed by the strand-filter\n",
      "1484 variant positions reported (1469 SNP, 15 indel)\n"
     ]
    }
   ],
   "source": [
    "!varscan mpileup2cns ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.mpileup --min-reads2 10 --output-vcf 1 > ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1eca74a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make an informed decision (based on section 1) about the parameters used for calling the variants.\n",
    "#There is no specific answer to this, but the idea is to look into the flags used, and make sure they make sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f15e953",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b9e448c4",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>2.2 Select any random variant in your VCF file which lies between the positions 20,800,000 and 30,800,000 on chromosome 16</h3>\n",
    "    <p><b>Q.</b> What is the average depth of bases for this variant called on chromosome 16</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "464292bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.3\n",
      "##source=VarScan2\n",
      "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 15\">\n",
      "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\n",
      "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\n",
      "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\n",
      "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\n",
      "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\n",
      "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
      "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 15\">\n",
      "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\n",
      "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\n",
      "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\n",
      "##FORMAT=<ID=PVAL,Number=1,Type=String,Description=\"P-value from Fisher's Exact Test\">\n",
      "##FORMAT=<ID=RBQ,Number=1,Type=Integer,Description=\"Average quality of reference-supporting bases (qual1)\">\n",
      "##FORMAT=<ID=ABQ,Number=1,Type=Integer,Description=\"Average quality of variant-supporting bases (qual2)\">\n",
      "##FORMAT=<ID=RDF,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on forward strand (reads1plus)\">\n",
      "##FORMAT=<ID=RDR,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on reverse strand (reads1minus)\">\n",
      "##FORMAT=<ID=ADF,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on forward strand (reads2plus)\">\n",
      "##FORMAT=<ID=ADR,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on reverse strand (reads2minus)\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSample1\n",
      "chr16\t29985888\t.\tC\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\n"
     ]
    }
   ],
   "source": [
    "#Extract this variant. There are multiple ways of doing this. Some fast and some easy.\n",
    "!grep -P '^(#|chr16\\s+29985888)' ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a6311294",
   "metadata": {},
   "outputs": [],
   "source": [
    "#There could be multiple variants in this range, please select one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "96c79a9f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#The information is there in your VCF file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0517cf21",
   "metadata": {},
   "source": [
    "The average depth of the variant is represented by ADP, which is one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea33031d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>2.3 How many indels do you have in your VCF file?</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0dab2c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Same as above, there are multiple ways of doing this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c3f951a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Write your command and show the number of indels."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7678ad4",
   "metadata": {},
   "source": [
    "There are 15 indels. We're still trying to figure it out."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e36f85aa",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>2.4 Select any random variant in your VCF file which lies between the position 203,000,000 and 230,000,000 on chromosome 2.</h3>\n",
    "    <p><b>2.4.1</b> What is the variation observed for your 1000 genome individual at this position? Is it 0/0 1/1/ 0/1 or 1/0? How many copies of reference alleles does your individual carry at this position?</p>\n",
    "    <p><b>2.4.2</b> How many read bases (forward and reverse) from your pileup file (as selected by VarScan) supported the alternative allele and how many variants supported the reference allele at this position?</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2d907cf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2\t219635909\t.\tG\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219635910\t.\tG\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219635911\t.\tG\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219636130\t.\tA\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219636131\t.\tT\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219636132\t.\tT\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219636133\t.\tG\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219636134\t.\tG\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219636135\t.\tG\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n",
      "chr2\t219636136\t.\tG\t.\t.\tPASS\tADP=1;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:1\r\n"
     ]
    }
   ],
   "source": [
    "!head ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf -n39000920 | tail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7613e4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dc62c42d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.3\n",
      "##source=VarScan2\n",
      "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 15\">\n",
      "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\n",
      "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\n",
      "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\n",
      "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\n",
      "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\n",
      "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
      "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 15\">\n",
      "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\n",
      "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\n",
      "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\n",
      "##FORMAT=<ID=PVAL,Number=1,Type=String,Description=\"P-value from Fisher's Exact Test\">\n",
      "##FORMAT=<ID=RBQ,Number=1,Type=Integer,Description=\"Average quality of reference-supporting bases (qual1)\">\n",
      "##FORMAT=<ID=ABQ,Number=1,Type=Integer,Description=\"Average quality of variant-supporting bases (qual2)\">\n",
      "##FORMAT=<ID=RDF,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on forward strand (reads1plus)\">\n",
      "##FORMAT=<ID=RDR,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on reverse strand (reads1minus)\">\n",
      "##FORMAT=<ID=ADF,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on forward strand (reads2plus)\">\n",
      "##FORMAT=<ID=ADR,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on reverse strand (reads2minus)\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSample1\n",
      "chr2\t219633235\t.\tA\t.\t.\tPASS\tADP=4;WT=0;HET=0;HOM=0;NC=1\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t./.:.:4\n"
     ]
    }
   ],
   "source": [
    "#Select the variant.\n",
    "!grep -P '^(#|chr2\\s+219633235)' ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6436c2db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr10' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270721v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr13' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_GL000009v2_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_GL000225v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270722v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_GL000194v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270723v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270724v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270725v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270726v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270727v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16_KI270728v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_GL000205v2_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270729v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr18' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270706v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270708v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270709v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270711v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270712v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270713v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270714v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr21' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270731v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270732v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270733v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270734v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270735v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270736v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270737v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270738v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_GL000221v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_GL000008v2_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_KI270717v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_KI270718v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_KI270719v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_KI270720v1_random' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270762v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270766v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270765v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_GL383518v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_GL383519v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_GL383520v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270764v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270763v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270759v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270761v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270770v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270773v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270774v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270769v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_GL383521v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270772v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270775v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270768v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_GL582966v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_GL383522v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270776v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr2_KI270767v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_JH636055v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270780v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_GL383526v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270777v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270779v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270782v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270784v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_KI270790v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_GL000257v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_KI270788v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_GL383527v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_KI270789v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_KI270793v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_KI270792v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_KI270791v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_GL949742v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_KI270794v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_GL339449v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_KI270795v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL000250v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_KI270800v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL383533v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_KI270801v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_KI270802v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_KB021644v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_KI270797v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_KI270798v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270804v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270809v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270806v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_GL383534v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270803v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270808v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270807v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270805v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270818v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270812v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270811v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270821v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270813v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270822v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270814v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270810v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270819v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270820v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270817v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270816v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270815v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_GL383539v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_GL383540v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_GL383541v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_GL383542v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9_KI270823v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr10_GL383545v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr10_KI270824v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr10_GL383546v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr10_KI270825v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270832v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270830v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270831v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270829v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_JH159136v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_JH159137v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270826v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_GL877875v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_GL877876v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_KI270837v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_GL383549v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_KI270835v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_GL383550v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_GL383552v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_GL383553v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_KI270834v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_KI270833v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr12_KI270836v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr13_KI270840v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr13_KI270839v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr13_KI270838v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr13_KI270842v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270844v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270847v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270845v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14_KI270846v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270852v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270851v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270848v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_GL383554v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270849v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_GL383555v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270850v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16_KI270854v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16_KI270856v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16_KI270855v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16_KI270853v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16_GL383556v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16_GL383557v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_GL383563v3_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270862v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270861v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270857v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_JH159146v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_JH159147v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_GL383564v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_GL000258v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_GL383565v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270858v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270859v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_GL383566v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270860v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr18_GL383567v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr18_GL383571v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr18_GL383572v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr18_KI270863v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270868v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270865v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL383573v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL383575v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL383576v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL383574v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270866v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270867v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949746v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr20_GL383577v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr20_KI270869v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr20_KI270871v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr20_KI270870v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr21_KI270874v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr21_KI270873v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr21_GL383579v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr21_GL383580v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr21_GL383581v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr21_KI270872v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270875v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270878v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270879v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270876v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270877v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_GL383583v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_GL383582v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrX_KI270880v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrX_KI270881v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270882v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270883v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270884v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270885v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr19_KI270886v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270887v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270888v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270889v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270890v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270891v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr1_KI270892v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270894v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr2_KI270893v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270895v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_KI270896v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_KI270897v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5_KI270898v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL000251v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7_KI270899v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270900v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270902v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270903v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12_KI270904v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270906v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15_KI270905v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270907v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270910v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270909v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_JH159148v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17_KI270908v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr18_KI270912v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr18_KI270911v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949747v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KB663609v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrX_KI270913v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270914v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270915v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270916v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270917v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270918v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270919v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270920v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270921v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270922v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270923v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270924v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4_KI270925v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL000252v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8_KI270926v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11_KI270927v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949748v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22_KI270928v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270929v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270930v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270931v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270932v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_KI270933v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL000209v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270934v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL000253v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949749v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270935v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL000254v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949750v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270936v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL000255v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949751v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3_KI270937v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_GL000256v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949752v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6_KI270758v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19_GL949753v2_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr19_KI270938v1_alt' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrM' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270442v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270467v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270438v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270512v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270519v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000195v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000219v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000220v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000224v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270741v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000213v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270743v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270744v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270745v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270746v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270747v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270748v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270749v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270750v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270751v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270753v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270756v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000214v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_KI270742v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000216v2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrUn_GL000218v1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrX' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrY' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    }
   ],
   "source": [
    "#filtering the file to only show variations\n",
    "!bcftools view -v snps,indels ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf > ~/biol6150/ProjectSubmissions/Group8/Project3/filtered_SRR393028.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4b7ddec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.3\r\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\r\n",
      "##source=VarScan2\r\n",
      "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 15\">\r\n",
      "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\r\n",
      "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\r\n",
      "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\r\n",
      "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\r\n",
      "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\r\n",
      "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\r\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\r\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\r\n",
      "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\r\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 15\">\r\n",
      "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\r\n",
      "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\r\n",
      "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\r\n",
      "##FORMAT=<ID=PVAL,Number=1,Type=String,Description=\"P-value from Fisher's Exact Test\">\r\n",
      "##FORMAT=<ID=RBQ,Number=1,Type=Integer,Description=\"Average quality of reference-supporting bases (qual1)\">\r\n",
      "##FORMAT=<ID=ABQ,Number=1,Type=Integer,Description=\"Average quality of variant-supporting bases (qual2)\">\r\n",
      "##FORMAT=<ID=RDF,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on forward strand (reads1plus)\">\r\n",
      "##FORMAT=<ID=RDR,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on reverse strand (reads1minus)\">\r\n",
      "##FORMAT=<ID=ADF,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on forward strand (reads2plus)\">\r\n",
      "##FORMAT=<ID=ADR,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on reverse strand (reads2minus)\">\r\n",
      "##bcftools_viewVersion=1.17+htslib-1.17\r\n",
      "##bcftools_viewCommand=view -v snps,indels /home/hice1/kbook6/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf; Date=Fri Oct 11 13:09:59 2024\r\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSample1\r\n",
      "chr1\t1955580\t.\tA\tG\t.\tPASS\tADP=14;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:76:14:14:0:14:100%:2.4927E-8:0:40:0:0:6:8\r\n",
      "chr1\t3823176\t.\tC\tT\t.\tPASS\tADP=22;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:45:22:22:10:12:54.55%:3.066E-5:33:32:2:8:5:7\r\n",
      "chr1\t4711993\t.\tT\tC\t.\tPASS\tADP=16;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:87:16:16:0:16:100%:1.6637E-9:0:39:0:0:0:16\r\n",
      "chr1\t7751155\t.\tA\tG\t.\tPASS\tADP=13;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:70:13:13:0:13:100%:9.6148E-8:0:32:0:0:11:2\r\n",
      "chr1\t9263943\t.\tA\tG\t.\tPASS\tADP=19;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:37:19:19:9:10:52.63%:1.9541E-4:36:38:6:3:6:4\r\n",
      "chr1\t9264611\t.\tA\tG\t.\tPASS\tADP=10;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:52:10:10:0:10:100%:5.4125E-6:0:29:0:0:5:5\r\n",
      "chr1\t11069791\t.\tA\tG\t.\tPASS\tADP=16;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:87:16:16:0:16:100%:1.6637E-9:0:30:0:0:0:16\r\n",
      "chr1\t11127836\t.\tT\tC\t.\tPASS\tADP=10;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:52:10:10:0:10:100%:5.4125E-6:0:32:0:0:2:8\r\n",
      "chr1\t12341811\t.\tA\tC\t.\tPASS\tADP=10;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:52:10:10:0:10:100%:5.4125E-6:0:33:0:0:5:5\r\n",
      "chr1\t12775725\t.\tC\tT\t.\tPASS\tADP=11;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:58:11:11:0:11:100%:1.4176E-6:0:33:0:0:2:9\r\n",
      "chr1\t12827690\t.\tC\tT\t.\tPASS\tADP=34;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:94:34:34:11:23:67.65%:3.5675E-10:30:32:4:7:11:12\r\n",
      "chr1\t12827720\t.\tT\tG\t.\tPASS\tADP=30;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:143:30:30:2:28:93.33%:4.194E-15:30:33:2:0:12:16\r\n",
      "chr1\t12838525\t.\tG\tC\t.\tPASS\tADP=13;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:70:13:13:0:13:100%:9.6148E-8:0:39:0:0:10:3\r\n",
      "chr1\t12838922\t.\tC\tG\t.\tPASS\tADP=18;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:76:18:18:2:16:88.89%:2.0936E-8:30:31:1:1:4:12\r\n",
      "chr1\t12838925\t.\tT\tA\t.\tPASS\tADP=17;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:81:17:17:1:16:94.12%:7.7134E-9:30:31:0:1:4:12\r\n",
      "chr1\t12882298\t.\tT\tC\t.\tPASS\tADP=24;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:135:24:24:0:24:100%:3.101E-14:0:30:0:0:7:17\r\n",
      "chr1\t12893335\t.\tT\tG\t.\tPASS\tADP=39;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:36:39:39:28:11:28.21%:2.1547E-4:35:32:17:11:5:6\r\n",
      "chr1\t12895136\t.\tT\tG\t.\tPASS\tADP=26;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:47:27:26:13:13:50%:1.6379E-5:32:34:3:10:5:8\r\n",
      "chr1\t12920302\t.\tC\tG\t.\tPASS\tADP=16;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:87:16:16:0:16:100%:1.6637E-9:0:33:0:0:3:13\r\n",
      "chr1\t13246759\t.\tT\tC\t.\tPASS\tADP=12;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:44:12:12:2:10:83.33%:3.3652E-5:45:33:2:0:4:6\r\n",
      "chr1\t13369423\t.\tC\tG\t.\tPASS\tADP=38;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:36:38:38:27:11:28.95%:2.1011E-4:31:38:15:12:5:6\r\n",
      "chr1\t13418237\t.\tC\tG\t.\tPASS\tADP=32;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:182:32:32:0:32:100%:5.4567E-19:0:30:0:0:18:14\r\n",
      "chr1\t15933897\t.\tA\tG\t.\tPASS\tADP=17;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:49:17:17:5:12:70.59%:1.1285E-5:36:31:2:3:6:6\r\n"
     ]
    }
   ],
   "source": [
    "#checking the head\n",
    "!head -50 ~/biol6150/ProjectSubmissions/Group8/Project3/filtered_SRR393028.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ba45ce2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2\t179945453\t.\tC\tT\t.\tPASS\tADP=10;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:52:10:10:0:10:100%:5.4125E-6:0:36:0:0:2:8\r\n",
      "chr2\t191386000\t.\tG\tA\t.\tPASS\tADP=18;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:99:18:18:0:18:100%:1.1019E-10:0:31:0:0:10:8\r\n",
      "chr2\t201258272\t.\tA\tG\t.\tPASS\tADP=15;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:81:15:15:0:15:100%:6.4467E-9:0:32:0:0:3:12\r\n",
      "chr2\t205743315\t.\tT\tC\t.\tPASS\tADP=11;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:58:12:11:0:11:100%:1.4176E-6:0:28:0:0:2:9\r\n",
      "chr2\t208171988\t.\tT\tC\t.\tPASS\tADP=15;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:81:15:15:0:15:100%:6.4467E-9:0:36:0:0:10:5\r\n",
      "chr2\t208172054\t.\tA\tG\t.\tPASS\tADP=21;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:117:21:21:0:21:100%:1.8578E-12:0:34:0:0:4:17\r\n",
      "chr2\t218454934\t.\tA\tG\t.\tPASS\tADP=11;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:58:11:11:0:11:100%:1.4176E-6:0:35:0:0:6:5\r\n",
      "chr2\t219175608\t.\tA\tG\t.\tPASS\tADP=16;WT=0;HET=1;HOM=0;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t0/1:44:16:16:5:11:68.75%:3.3854E-5:30:33:3:2:3:8\r\n",
      "chr2\t219213774\t.\tT\tC\t.\tPASS\tADP=14;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:76:14:14:0:14:100%:2.4927E-8:0:38:0:0:13:1\r\n",
      "chr2\t219224963\t.\tG\tC\t.\tPASS\tADP=10;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:52:10:10:0:10:100%:5.4125E-6:0:30:0:0:8:2\r\n"
     ]
    }
   ],
   "source": [
    "#looking from the tail for chr2\n",
    "!head ~/biol6150/ProjectSubmissions/Group8/Project3/filtered_SRR393028.vcf -n900 | tail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "15b2f057",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.3\r\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\r\n",
      "##source=VarScan2\r\n",
      "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 15\">\r\n",
      "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\r\n",
      "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\r\n",
      "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\r\n",
      "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\r\n",
      "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\r\n",
      "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\r\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\r\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\r\n",
      "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\r\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 15\">\r\n",
      "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\r\n",
      "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\r\n",
      "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\r\n",
      "##FORMAT=<ID=PVAL,Number=1,Type=String,Description=\"P-value from Fisher's Exact Test\">\r\n",
      "##FORMAT=<ID=RBQ,Number=1,Type=Integer,Description=\"Average quality of reference-supporting bases (qual1)\">\r\n",
      "##FORMAT=<ID=ABQ,Number=1,Type=Integer,Description=\"Average quality of variant-supporting bases (qual2)\">\r\n",
      "##FORMAT=<ID=RDF,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on forward strand (reads1plus)\">\r\n",
      "##FORMAT=<ID=RDR,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on reverse strand (reads1minus)\">\r\n",
      "##FORMAT=<ID=ADF,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on forward strand (reads2plus)\">\r\n",
      "##FORMAT=<ID=ADR,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on reverse strand (reads2minus)\">\r\n",
      "##bcftools_viewVersion=1.17+htslib-1.17\r\n",
      "##bcftools_viewCommand=view -v snps,indels /home/hice1/kbook6/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf; Date=Fri Oct 11 13:09:59 2024\r\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSample1\r\n",
      "chr2\t219224963\t.\tG\tC\t.\tPASS\tADP=10;WT=0;HET=0;HOM=1;NC=0\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:52:10:10:0:10:100%:5.4125E-6:0:30:0:0:8:2\r\n"
     ]
    }
   ],
   "source": [
    "#grep for the specific chromosome 2 variant 219224963 \n",
    "!grep -P '^(#|chr2\\s+219224963)' ~/biol6150/ProjectSubmissions/Group8/Project3/filtered_SRR393028.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8f8dcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#looking for the variant **IGNORE**\n",
    "#!grep -v '^#' ~/biol6150/ProjectSubmissions/Group8/Project3/VariantCalling/SRR393028.vcf | awk '{if(length($4) == 1 && length($5) == 1 || length($4) > 1 || length($5) > 1) print $0}' > variants_snps_indels.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f2f3591f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Variation observed and number of copies for reference allele."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f60273e",
   "metadata": {},
   "source": [
    "Variation position: Chr2 219334963"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "403ccc29",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number of reads supporting the ref & alt allele."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "064ae9d3",
   "metadata": {},
   "source": [
    "Number of copies: Since this variant is 1/1, it is homozygous alternate. Therefore, there are 2 copes of the alternate alleles and 0 copies of the reference allele. It is also 0 since wt=0, which is the number of samples called reference \"wild type.\"\n",
    "*This is an unphased file, so the 1 and 0 are interchangable.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9841242e",
   "metadata": {},
   "source": [
    "# **3. Filtering**\n",
    "### *Total Questions: 1*\n",
    "### *Total Points: 8 = 8*\n",
    "\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63f0b0dc",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <h3>3.1 Extract variants that have average per sample depth of > 50. </h3>\n",
    "    <p>Show your command and number of variants before and after</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "61cd2df9",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Min coverage:\t10\n",
      "Min reads2:\t50\n",
      "Min strands2:\t1\n",
      "Min var freq:\t0.2\n",
      "Min avg qual:\t15\n",
      "P-value thresh:\t0.1\n",
      "Reading input from /home/hice1/kbook6/biol6150/ProjectSubmissions/Group8/Project3/filtered_SRR393028.vcf\n",
      "1484 entries in input stream\n",
      "0 had no genotype\n",
      "0 were called wild-type\n",
      "0 failed coverage\n",
      "1476 failed reads2\n",
      "0 failed strands\n",
      "0 failed quality\n",
      "0 failed variant frequency < 0.2\n",
      "0 failed P-value > 0.1\n",
      "0 were removed near indels\n",
      "8 passed filters\n"
     ]
    }
   ],
   "source": [
    "#finding variants with at least 25 reads supporting the alt allele\n",
    "!varscan filter ~/biol6150/ProjectSubmissions/Group8/Project3/filtered_SRR393028.vcf --min-reads2 50 --output-file ~/biol6150/ProjectSubmissions/Group8/Project3/VSfiltered_SRR393028.vcf "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a59ffd1",
   "metadata": {},
   "source": [
    "Originally, 1484 variants were enterd in the command, and 1476 failed to pass the filters. \n",
    "8 reads passed the filters, meaning that 8 variants have a depth of at least 50."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:biol6150]",
   "language": "python",
   "name": "conda-env-biol6150-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
